<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Generalized Robinson-Foulds distances • TreeDist</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Generalized Robinson-Foulds distances">
<meta name="robots" content="noindex">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-BTKEG66D3F"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-BTKEG66D3F');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TreeDist</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">2.9.2.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/Using-TreeDist.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Function reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tree distance analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/Using-TreeDist.html">Calculate tree similarity with 'TreeDist'</a></li>
    <li><a class="dropdown-item" href="../articles/using-distances.html">Contextualizing tree distances</a></li>
    <li><a class="dropdown-item" href="../articles/different-leaves.html">Trees with different leaves</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tree space analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/treespace.html">Tree space analysis</a></li>
    <li><a class="dropdown-item" href="../articles/landscapes.html">Analysing landscapes of phylogenetic trees</a></li>
    <li><a class="dropdown-item" href="../articles/compare-treesets.html">Comparing sets of trees from different analyses</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tree distance introductions</h6></li>
    <li><a class="dropdown-item" href="../articles/information.html">Comparing splits using information theory</a></li>
    <li><a class="dropdown-item" href="../articles/Robinson-Foulds.html">Extending the Robinson-Foulds metric</a></li>
    <li><a class="dropdown-item" href="../articles/Generalized-RF.html">Generalized Robinson-Foulds distances</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ms609/TreeDist"><span class="fa fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Generalized Robinson-Foulds distances</h1>
                        <h4 data-toc-skip class="author"><a href="https://smithlabdurham.github.io/" class="external-link">Martin R. Smith</a></h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/ms609/TreeDist/blob/main/vignettes/Generalized-RF.Rmd" class="external-link"><code>vignettes/Generalized-RF.Rmd</code></a></small>
      <div class="d-none name"><code>Generalized-RF.Rmd</code></div>
    </div>

    
    
<p>This document outlines the similarity measures employed by the
generalized Robinson–Foulds distances implemented in this package.</p>
<p>Generalized RF distances are introduced <a href="Robinson-Foulds.html">elsewhere</a>; before you read further, you
may also wish to revisit how to <a href="Using-TreeDist.html">use the
‘TreeDist’ package</a>, and relevant <a href="information.html">principles of information theory</a>.</p>
<div class="section level2">
<h2 id="shared-phylogenetic-information">Shared phylogenetic information<a class="anchor" aria-label="anchor" href="#shared-phylogenetic-information"></a>
</h2>
<p>Under the shared phylogenetic information tree distance measure <span class="citation">(Smith, 2020)</span>, pairs of splits are assigned a
similarity score that corresponds to the amount of phylogenetic
information <span class="citation">(<em>sensu</em> Steel &amp; Penny,
2006)</span> that they share in common (see <a href="information.html">separate vignette</a>), a concept introduced
(though not developed) by Nelson <span class="citation">(1979)</span>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">SharedPhylogeneticInfo</span>, <span class="va">tree1</span>, <span class="va">tree2</span>, </span>
<span>                  Plot <span class="op">=</span> <span class="va">TreeDistPlot</span>, matchZeros <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-1-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/TreeDistance.html">SharedPhylogeneticInfo</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 12.32188</span></span></code></pre>
<p>This distance is measured in bits; on this measure, the total
information content of a tree is given by</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/TreeInfo.html">SplitwiseInfo</a></span><span class="op">(</span><span class="va">tree1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 28.67404</span></span></code></pre>
<div class="section level3">
<h3 id="conflicting-splits-can-nevertheless-be-instructive">Conflicting splits can nevertheless be instructive<a class="anchor" aria-label="anchor" href="#conflicting-splits-can-nevertheless-be-instructive"></a>
</h3>
<p>Shared phylogenetic information assigns zero similarity to
incompatible splits, i.e. those that cannot both occur on a single tree.
This leads to problematic behaviour in certain cases: for example,
swapping the position of two distant leaves (as with leaves ‘A’ and ‘J’
below) can disproportionately reduce similarity – in this example, to
zero.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">SharedPhylogeneticInfo</span>, <span class="va">AtoJ</span>, <span class="va">swapAJ</span>,</span>
<span>                  Plot <span class="op">=</span> <span class="va">TreeDistPlot</span>, matchZeros <span class="op">=</span> <span class="cn">FALSE</span>, prune <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">18</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-2-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="mutual-clustering-information">Mutual clustering information<a class="anchor" aria-label="anchor" href="#mutual-clustering-information"></a>
</h2>
<p>Scoring each pair of splits according to their mutual clustering
information <span class="citation">(Smith, 2020)</span> (see <a href="information.html">separate vignette</a>) results in a
information-based tree distance metric that recognizes similarity in
tree structure even when every possible pairing of splits conflicts:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">MutualClusteringInfo</span>, <span class="va">AtoJ</span>, <span class="va">swapAJ</span>,</span>
<span>                  Plot <span class="op">=</span> <span class="va">TreeDistPlot</span>, matchZeros <span class="op">=</span> <span class="cn">FALSE</span>, prune <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">18</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-3-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/TreeDistance.html">MutualClusteringInfo</a></span><span class="op">(</span><span class="va">AtoJ</span>, <span class="va">swapAJ</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1.348339</span></span></code></pre>
<p>Because no pair of non-trivial splits has zero mutual clustering
information, even a dissimilar pairing (such as <code>HI|ABCDEFG</code>
⇒ <code>EI|ABCDFGH</code> below) is (slightly) preferable to leaving a
split unpaired.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">MutualClusteringInfo</span>, <span class="va">tree1</span>, <span class="va">tree2</span>, </span>
<span>                  Plot <span class="op">=</span> <span class="va">TreeDistPlot</span>, matchZeros <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The total mutual clustering information in a single tree is given
by</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/TreeInfo.html">ClusteringEntropy</a></span><span class="op">(</span><span class="va">tree1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 5.120281</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="nye-et-al--tree-similarity-metric">Nye <em>et al.</em> tree similarity metric<a class="anchor" aria-label="anchor" href="#nye-et-al--tree-similarity-metric"></a>
</h2>
<p>The Nye <em>et al</em>. <span class="citation">(2006)</span> tree
similarity metric scores pairs by considering the elements held in
common between subsets of each split.</p>
<p>Consider a pair of splits <code>ABCDEF|GHIJ</code> and
<code>ABCDEIJ|FGH</code>. These can be aligned thus:</p>
<pre><code>ABCDEF  | GHIJ
ABCDE IJ|FGH</code></pre>
<p>The first pair of subsets, <code>ABCDEF</code> and
<code>ABCDEIJ</code>, have five elements in common (<code>ABCDE</code>),
and together encompass eight elements (<code>ABCDEFIJ</code>). Their
<em>subset score</em> is thus
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mn>5</mn><mn>8</mn></mfrac><annotation encoding="application/x-tex">\frac{5}{8}</annotation></semantics></math>.</p>
<p>The second pair of subsets, <code>GHIJ</code> and <code>FGH</code>,
have two elements (<code>GH</code>) in common, of the five total
(<code>FGHIJ</code>), and hence receive a subset score of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mn>2</mn><mn>5</mn></mfrac><annotation encoding="application/x-tex">\frac{2}{5}</annotation></semantics></math>.</p>
<p>This split alignment then receives an <em>alignment score</em>
corresponding to the lower of the two subset scores,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mn>2</mn><mn>5</mn></mfrac><annotation encoding="application/x-tex">\frac{2}{5}</annotation></semantics></math>.</p>
<p>We must now consider the other alignment of this pair of splits,</p>
<pre><code>ABCDEF  |     GHIJ
     FGH|ABCDE  IJ</code></pre>
<p>This yields subset scores of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mn>1</mn><mn>8</mn></mfrac><annotation encoding="application/x-tex">\frac{1}{8}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mn>2</mn><mn>9</mn></mfrac><annotation encoding="application/x-tex">\frac{2}{9}</annotation></semantics></math>,
and thus has an alignment score of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mn>1</mn><mn>8</mn></mfrac><annotation encoding="application/x-tex">\frac{1}{8}</annotation></semantics></math>.
This alignment gives a lower score than the other, so is disregarded.
The pair of splits is allocated a similarity score corresponding to the
better alignment:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mn>2</mn><mn>5</mn></mfrac><annotation encoding="application/x-tex">\frac{2}{5}</annotation></semantics></math>.</p>
<p>As such, splits that match exactly will receive a similarity score of
1, in a manner analogous to the Robinson–Foulds distance. (This is
despite the fact that some splits are more likely to match than others.)
It is not possible for a pair of splits to receive a zero similarity
score.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">NyeSimilarity</span>, <span class="va">tree1</span>, <span class="va">tree2</span>, </span>
<span>                  Plot <span class="op">=</span> <span class="va">TreeDistPlot</span>, matchZeros <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-5-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/NyeSimilarity.html">NyeSimilarity</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>, normalize <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 3.5</span></span></code></pre>
<div class="section level3">
<h3 id="jaccardrobinsonfoulds-metric">Jaccard–Robinson–Foulds metric<a class="anchor" aria-label="anchor" href="#jaccardrobinsonfoulds-metric"></a>
</h3>
<p>Böcker <em>et al</em>. <span class="citation">(2013)</span> employ
the same split similarity calculation as Nye <em>et al.</em> (above),
which they suggest ought to be raised to an arbitrary exponent in order
to down-weight the contribution of paired splits that are not identical.
In order for the metric to converge to the Robinson–Foulds metric as the
exponent grows towards infinity, the resulting score is then
doubled.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/JaccardRobinsonFoulds.html">JaccardRobinsonFoulds</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>, k <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 4</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">JaccardRobinsonFoulds</span>, <span class="va">tree1</span>, <span class="va">tree2</span>,</span>
<span>                  Plot <span class="op">=</span> <span class="va">TreeDistPlot</span>, matchZeros <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">JRF2</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">...</span><span class="op">)</span> <span class="fu"><a href="../reference/JaccardRobinsonFoulds.html">JaccardRobinsonFoulds</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">2</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="fu">JRF2</span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 5.638889</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">JRF2</span>, <span class="va">tree1</span>, <span class="va">tree2</span>,</span>
<span>                  Plot <span class="op">=</span> <span class="va">TreeDistPlot</span>, matchZeros <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-6-2.png" width="576" style="display: block; margin: auto;"></p>
<p>The figure below shows how the JRF distance between the two trees
plotted above varies with the value of the exponent <em>k</em>, relative
to the Nye <em>et al.</em> and Robinson–Foulds distances between the
trees:</p>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-7-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The theoretical and practical performance of the JRF metric, and its
speed of calculation, are best at lower values of <em>k</em> <span class="citation">(Smith, 2020)</span>, raising the question of whether
an exponent is useful.</p>
<p>Böcker <em>et al</em>. <span class="citation">(2013)</span> suggest
that ‘reasonable’ matchings exhibit a property they term
<em>arboreality</em>. Their definition of an arboreal matching supposes
that trees are rooted, in which case each split corresponds to a clade.
In an arboreal matching on a rooted tree, no pairing of splits conflicts
with any other pairing of splits. Consider the case where splits ‘A’ and
‘B’ in tree 1 are paired respectively with splits ‘C’ and ‘D’ in tree 2.
If ‘A’ is paired with ‘C’ and ‘B’ with ‘D’, then to avoid conflict:</p>
<ul>
<li><p>if A is nested within B, then C must be nested within D</p></li>
<li><p>if B is nested within A, then D must be nested within C</p></li>
<li><p>if A and B do not overlap, then C and D must not
overlap.</p></li>
</ul>
<p>Equivalent statements for unrooted trees are a little harder to
express.</p>
<p>Unfortunately, constructing arboreal matchings is NP-complete, making
an optimal arboreal matching slow to find. A faster alternative is to
prohibit pairings of contradictory splits, though distances generated
under such an approach are theoretically less coherent and practically
no more effective than those calculated when contradictory splits may be
paired, so the only advantage of this approach is a slight increase in
calculation speed <span class="citation">(Smith, 2020)</span>.</p>
</div>
</div>
<div class="section level2">
<h2 id="matching-split-distance">Matching Split Distance<a class="anchor" aria-label="anchor" href="#matching-split-distance"></a>
</h2>
<p>Bogdanowicz &amp; Giaro <span class="citation">(2012)</span> propose
an alternative distance, which they term the Matching Split Distance.
(This approach was independently proposed by Lin <em>et al.</em> <span class="citation">(2012)</span>.)</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MatchingSplitDistance.html">MatchingSplitDistance</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 5</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">MatchingSplitDistance</span>, <span class="va">tree1</span>, <span class="va">tree2</span>,</span>
<span>                  Plot <span class="op">=</span> <span class="va">TreeDistPlot</span>, matchZeros <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Note that the visualization shows the difference, rather than the
similarity, between splits.</p>
<p>Similar to the Nye <em>et al</em>. similarity metric, this method
compares the subsets implied by a pair of splits. Here, the relevant
quantity is the number of elements that must be moved from one subset to
another in order to make the two splits identical. With the pair of
splits</p>
<pre><code>ABCDEF  | GHIJ
ABCDE IJ|FGH</code></pre>
<p>three leaves (‘F’, ‘I’ and ‘J’) must be moved before the splits are
identical; as such, the pair of splits are assigned a difference score
of three.<br>
Formally, where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>S</mi><mi>i</mi></msub><annotation encoding="application/x-tex">S_i</annotation></semantics></math>
splits
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
leaves into bipartitions
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>A</mi><mi>i</mi></msub><annotation encoding="application/x-tex">A_i</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>B</mi><mi>i</mi></msub><annotation encoding="application/x-tex">B_i</annotation></semantics></math>,
the difference score is calculated by</p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>−</mo><mi>m</mi></mrow><annotation encoding="application/x-tex">n - m</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>
counts the number of leaves that already match, and is defined as</p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>m</mi><mo>=</mo><mo>max</mo><mo stretchy="false" form="prefix">{</mo><mrow><mo stretchy="true" form="prefix">|</mo><msub><mi>A</mi><mn>1</mn></msub><mo>∩</mo><msub><mi>A</mi><mn>2</mn></msub><mo stretchy="true" form="postfix">|</mo></mrow><mo>+</mo><mrow><mo stretchy="true" form="prefix">|</mo><msub><mi>B</mi><mn>1</mn></msub><mo>∩</mo><msub><mi>B</mi><mn>2</mn></msub><mo stretchy="true" form="postfix">|</mo></mrow><mo>,</mo><mrow><mo stretchy="true" form="prefix">|</mo><msub><mi>A</mi><mn>1</mn></msub><mo>∩</mo><msub><mi>B</mi><mn>2</mn></msub><mo stretchy="true" form="postfix">|</mo></mrow><mo>+</mo><mrow><mo stretchy="true" form="prefix">|</mo><msub><mi>B</mi><mn>1</mn></msub><mo>∩</mo><msub><mi>A</mi><mn>2</mn></msub><mo stretchy="true" form="postfix">|</mo></mrow><mo stretchy="false" form="postfix">}</mo></mrow><annotation encoding="application/x-tex">m = \max\{|A_1 \cap A_2| + |B_1 \cap B_2|, |A_1 \cap B_2| + |B_1 \cap A_2|\}</annotation></semantics></math></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MatchingSplitDistance.html">MatchingSplitDistance</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span>text<span class="op">=</span><span class="st">'((a, b, c, d, e, f), (g, h, i, j));'</span><span class="op">)</span>,</span>
<span>                      <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span>text<span class="op">=</span><span class="st">'((a, b, c, d, e, i, j), (g, h, f));'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 3</span></span></code></pre>
<p>This distance is difficult to normalize, as it is not easy to
calculate its maximum value.</p>
<div class="section level3">
<h3 id="information-theoretic-alternative">Information theoretic alternative<a class="anchor" aria-label="anchor" href="#information-theoretic-alternative"></a>
</h3>
<p>In the matching split distance,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>
represents a simple count of the number of shared taxa. An alternative
is to measure the phylogenetic information content of the largest split
consistent with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>S</mi><mn>1</mn></msub><annotation encoding="application/x-tex">S_1</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>S</mi><mn>2</mn></msub><annotation encoding="application/x-tex">S_2</annotation></semantics></math>:</p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>m</mi><mo>=</mo><mo>max</mo><mo stretchy="false" form="prefix">{</mo><mi>h</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>A</mi><mn>1</mn></msub><mo>∩</mo><msub><mi>A</mi><mn>2</mn></msub><mo stretchy="false" form="prefix">|</mo><msub><mi>B</mi><mn>1</mn></msub><mo>∩</mo><msub><mi>B</mi><mn>2</mn></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo><mi>h</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>A</mi><mn>1</mn></msub><mo>∩</mo><msub><mi>B</mi><mn>2</mn></msub><mo stretchy="false" form="prefix">|</mo><msub><mi>B</mi><mn>1</mn></msub><mo>∩</mo><msub><mi>A</mi><mn>2</mn></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="false" form="postfix">}</mo></mrow><annotation encoding="application/x-tex">m = \max\{h(A_1 \cap A_2 | B_1 \cap B_2), h(A_1 \cap B_2 | B_1 \cap A_2)\}</annotation></semantics></math></p>
<p>The most information-rich split consistent with</p>
<pre><code>ABCDEF  | GHIJ
ABCDE IJ|FGH  </code></pre>
<p>is <code>ABCDE | GH</code>, which contains</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">TreeTools</span><span class="fu">::</span><span class="fu"><a href="https://ms609.github.io/TreeTools/reference/SplitInformation.html" class="external-link">SplitInformation</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 3.169925</span></span></code></pre>
<p>bits of phylogenetic information. This value can be used as a
similarity score for this pairing of splits.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/TreeDistance.html">MatchingSplitInfoDistance</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 17.27586</span></span></code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">MatchingSplitInfoDistance</span>, <span class="va">tree1</span>, <span class="va">tree2</span>,</span>
<span>                  Plot <span class="op">=</span> <span class="va">TreeDistPlot</span>, matchZeros <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-11-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0" line-spacing="2">
<div id="ref-Bocker2013" class="csl-entry">
Böcker, S., Canzar, S., &amp; Klau, G. W. (2013). The generalized
<span>Robinson</span>-<span>Foulds</span> metric. In A. Darling &amp; J.
Stoye (Eds.), <em>Algorithms in <span>Bioinformatics</span>.
<span>WABI</span> 2013. <span>Lecture</span> <span>Notes</span> in
<span>Computer</span> <span>Science</span>, vol 8126</em> (pp. 156–169).
Berlin, Heidelberg: Springer. doi: <a href="https://doi.org/10.1007/978-3-642-40453-5_13" class="external-link">10.1007/978-3-642-40453-5_13</a>
</div>
<div id="ref-Bogdanowicz2012" class="csl-entry">
Bogdanowicz, D., &amp; Giaro, K. (2012). <span class="nocase">Matching
split distance for unrooted binary phylogenetic trees</span>.
<em>IEEE/ACM Transactions on Computational Biology and
Bioinformatics</em>, <em>9</em>(1), 150–160. doi: <a href="https://doi.org/10.1109/TCBB.2011.48" class="external-link">10.1109/TCBB.2011.48</a>
</div>
<div id="ref-Lin2012" class="csl-entry">
Lin, Y., Rajan, V., &amp; Moret, B. M. E. (2012). A metric for
phylogenetic trees based on matching. <em>IEEE/ACM Transactions on
Computational Biology and Bioinformatics</em>, <em>4</em>(9), 1014–1022.
doi: <a href="https://doi.org/10.1109/TCBB.2011.157" class="external-link">10.1109/TCBB.2011.157</a>
</div>
<div id="ref-Nelson1979" class="csl-entry">
Nelson, G. (1979). Cladistic analysis and synthesis: Principles and
definitions, with a historical note on <span>Adanson</span>’s
<em><span>Familles</span> des <span>Plantes</span></em> (1763–1764).
<em>Systematic Biology</em>, <em>28</em>(1), 1–21. doi: <a href="https://doi.org/10.1093/sysbio/28.1.1" class="external-link">10.1093/sysbio/28.1.1</a>
</div>
<div id="ref-Nye2006" class="csl-entry">
Nye, T. M. W., Liò, P., &amp; Gilks, W. R. (2006). <span class="nocase">A novel algorithm and web-based tool for comparing two
alternative phylogenetic trees</span>. <em>Bioinformatics</em>,
<em>22</em>(1), 117–119. doi: <a href="https://doi.org/10.1093/bioinformatics/bti720" class="external-link">10.1093/bioinformatics/bti720</a>
</div>
<div id="ref-SmithDist" class="csl-entry">
Smith, M. R. (2020). Information theoretic <span>Generalized
Robinson-Foulds</span> metrics for comparing phylogenetic trees.
<em>Bioinformatics</em>, <em>36</em>(20), 5007–5013. doi: <a href="https://doi.org/10.1093/bioinformatics/btaa614" class="external-link">10.1093/bioinformatics/btaa614</a>
</div>
<div id="ref-Steel2006" class="csl-entry">
Steel, M. A., &amp; Penny, D. (2006). Maximum parsimony and the
phylogenetic information in multistate characters. In V. A. Albert
(Ed.), <em>Parsimony, phylogeny, and genomics</em> (pp. 163–178).
Oxford: Oxford University Press.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://smithlabdurham.github.io/" class="external-link">Martin R. Smith</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.9000.</p>
</div>

    </footer>
</div>





  </body>
</html>
