<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Extending the Robinson-Foulds metric • TreeDist</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Extending the Robinson-Foulds metric">
<meta property="og:description" content="TreeDist">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-156666971-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156666971-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TreeDist</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="In-development version">1.1.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/Using-TreeDist.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Using-TreeDist.html">Using TreeDist to calculate tree similarity</a>
    </li>
    <li>
      <a href="../articles/using-distances.html">Contextualizing tree distances</a>
    </li>
    <li>
      <a href="../articles/Robinson-Foulds.html">Robinson-Foulds metric</a>
    </li>
    <li>
      <a href="../articles/Generalized-RF.html">Generalized Robinson-Foulds distances</a>
    </li>
    <li>
      <a href="../articles/information.html">Concepts of information</a>
    </li>
    <li>
      <a href="../articles/Tree-distance-interpretation.html">Evaluating tree distance metrics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ms609/TreeDist">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Extending the Robinson-Foulds metric</h1>
                        <h4 class="author">Martin R. Smith</h4>
            
            <h4 class="date">2020-07-08</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ms609/TreeDist/blob/master/vignettes/Robinson-Foulds.Rmd"><code>vignettes/Robinson-Foulds.Rmd</code></a></small>
      <div class="hidden name"><code>Robinson-Foulds.Rmd</code></div>

    </div>

    
    
<div id="the-robinson-foulds-distance" class="section level1">
<h1 class="hasAnchor">
<a href="#the-robinson-foulds-distance" class="anchor"></a>The Robinson-Foulds distance</h1>
<p>An intuitive way to calculate the similarity between two trees is to calculate the resolution of their strict consensus, which corresponds to the number of splits that occur in both trees <span class="citation">(Schuh &amp; Polhemus, 1980)</span>. The corresponding distance measure, the Robinson-Foulds distance <span class="citation">(Robinson &amp; Foulds, 1981)</span>, counts the number of splits that are unique to one of the two trees.</p>
<p>It is important to remember that counting splits is not the same as counting clades, edges or nodes. If a tree is drawn as rooted (even without a root edge), then the number of splits is one less than the number of edges or clades, and two less than the number of nodes. For example, the trees below have one split, two edges and three nodes in common.</p>
<p><img src="Robinson-Foulds_files/figure-html/unnamed-chunk-1-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The simplicity of counting splits is appealing, but limited by the underlying assumption that all splits are equivalent.</p>
<p>As an example, a split that separates eight leaves into two sets of four (as in the right-hand tree above) has a <span class="math inline">\(\frac{1}{35}\)</span> chance of being compatible with the reference tree. In contrast, a split that separates two leaves from the other six has a <span class="math inline">\(\frac{1}{7}\)</span> chance of matching the reference tree: the similarity observed is five times more likely to have arisen by chance. In other words, failure to match an even split is less noteworthy than failure to match an uneven one.</p>
<p>As a consequence, trees whose splits are less even will, on average, exhibit higher Robinson-Foulds distances with comparison trees. Compare a balanced and an unbalanced eight-taxon tree:</p>
<p><img src="Robinson-Foulds_files/figure-html/two-trees-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Each tree divides the eight taxa into five splits. The <a href="information.html">phylogenetic information content</a> of a split is a function of the probability that the split will match a uniformly chosen random tree, i.e. the proportion of eight-leaf binary trees that contain the split in question. (Information content, in bits, is defined as <span class="math inline">\(-\log_2(\textrm{probability})\)</span>.) This, in turn, is a function of the evenness of the split:</p>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="left">Matching trees</th>
<th align="left">
<em>P</em>(Match in random tree)</th>
<th align="left">Phylogenetic information content</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Split size: 2:6</td>
<td align="left">945 / 10 395</td>
<td align="left">0.0909</td>
<td align="left">3.46 bits</td>
</tr>
<tr class="even">
<td align="left">Split size: 3:5</td>
<td align="left">315 / 10 395</td>
<td align="left">0.0303</td>
<td align="left">5.04 bits</td>
</tr>
<tr class="odd">
<td align="left">Split size: 4:4</td>
<td align="left">225 / 10 395</td>
<td align="left">0.0216</td>
<td align="left">5.53 bits</td>
</tr>
</tbody>
</table>
<p>In the first tree, split 1 is even, dividing four taxa from four others (<code>4|4</code>); splits 2–5 are maximally uneven (<code>2|6</code>). If each split is treated as independent, then the total information content of the five splits is 19.37 bits, whereas that of the five splits in the second tree, of sizes <code>2|6</code>, <code>3|5</code>, <code>4|4</code>, <code>3|5</code> and <code>2|6</code>, is 22.54 bits. Put another way, a random tree will on average share more splits with the balanced tree (whose splits are predominantly uneven and thus likely to be matched) than the asymmetric tree (which contains more even splits that are less likely to occur in a random tree).</p>
<p>Indeed, of the 10 395 eight-leaf trees, many more bear at least one split in common with a balanced tree than with an asymmetric tree:</p>
<p><img src="Robinson-Foulds_files/figure-html/all-8-tip-trees-1.png" width="384" style="display: block; margin: auto;"></p>
</div>
<div id="information-corrected-robinson-foulds-distance" class="section level1">
<h1 class="hasAnchor">
<a href="#information-corrected-robinson-foulds-distance" class="anchor"></a>Information-corrected Robinson-Foulds distance</h1>
<p>This differing information content can be accommodated by weighting each split according to the amount of phylogenetic information it contains <span class="citation">(Smith, 2020)</span>. The two tree pairs below both have a Robinson-Foulds distance of two, but the first pair differ with regard to an uneven split (<code>ABCDEF|GH</code>), so obtain a total difference of 22.54 − (3.46 + 5.04 + 5.53 + 5.04) = 3.46 bits:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="no">tree1</span> <span class="kw">&lt;-</span> <span class="kw pkg">ape</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span>(<span class="kw">text</span><span class="kw">=</span><span class="st">'(1, (2, (3, (4, (5, (6, (7, 8)))))));'</span>)
<span class="no">tree2</span> <span class="kw">&lt;-</span> <span class="kw pkg">ape</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span>(<span class="kw">text</span><span class="kw">=</span><span class="st">'(1, (2, (3, (4, (5, (7, (6, 8)))))));'</span>)
<span class="no">tree3</span> <span class="kw">&lt;-</span> <span class="kw pkg">ape</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span>(<span class="kw">text</span><span class="kw">=</span><span class="st">'(1, (2, (3, (5, (4, (6, (7, 8)))))));'</span>)

<span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(<span class="no">InfoRobinsonFoulds</span>, <span class="no">tree1</span>, <span class="no">tree2</span>,
                  <span class="kw">Plot</span> <span class="kw">=</span> <span class="no">TreeDistPlot</span>, <span class="kw">prune</span> <span class="kw">=</span> <span class="fl">12</span>)</pre></body></html></div>
<p><img src="Robinson-Foulds_files/figure-html/unnamed-chunk-2-1.png" width="576" style="display: block; margin: auto;"></p>
<p>whereas the second pair differ in the resolution of a more even, and thus more information-rich, split (<code>ABCD|EFGH</code>), and so receive a distance score of 5.53 bits:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(<span class="no">InfoRobinsonFoulds</span>, <span class="no">tree1</span>, <span class="no">tree3</span>,
                  <span class="kw">Plot</span> <span class="kw">=</span> <span class="no">TreeDistPlot</span>, <span class="kw">prune</span> <span class="kw">=</span> <span class="fl">8</span>)</pre></body></html></div>
<p><img src="Robinson-Foulds_files/figure-html/unnamed-chunk-3-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="generalized-robinson-foulds-distances" class="section level1">
<h1 class="hasAnchor">
<a href="#generalized-robinson-foulds-distances" class="anchor"></a>Generalized Robinson-Foulds distances</h1>
<p>Even when accounting for the information content of splits in this way, the Robinson-Foulds distance is readily saturated: the maximum value can be obtained by moving a single leaf.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">tree1</span> <span class="kw">&lt;-</span> <span class="kw pkg">ape</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span>(<span class="kw">text</span><span class="kw">=</span><span class="st">'(1, (2, (3, (4, (5, (6, (7, 8)))))));'</span>)
<span class="no">tree2</span> <span class="kw">&lt;-</span> <span class="kw pkg">ape</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span>(<span class="kw">text</span><span class="kw">=</span><span class="st">'(8, (1, (2, (3, (4, (5, (6, 7)))))));'</span>)

<span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(<span class="no">RobinsonFouldsMatching</span>, <span class="no">tree1</span>, <span class="no">tree2</span>, <span class="kw">Plot</span> <span class="kw">=</span> <span class="no">TreeDistPlot</span>)</pre></body></html></div>
<p><img src="Robinson-Foulds_files/figure-html/unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Generalized Robinson-Foulds distances <span class="citation">(Böcker, Canzar, &amp; Klau, 2013; Nye, Liò, &amp; Gilks, 2006)</span> seek to address this issue. This category of metrics aim to acknowledge semblances between similar-but-not-quite-identical pairs of splits, which would contribute zero to tree similarity under the standard Robinson-Foulds measure.</p>
<p>Generalized RF distances work by finding a <em>matching</em> that pairs splits from one tree with splits in the other. Each pairing is scored according to the similarity of the paired splits; the sum of these scores is the score of the matching. The tree distance is given by the score of the optimal matching.</p>
<div id="constructing-a-matching" class="section level2">
<h2 class="hasAnchor">
<a href="#constructing-a-matching" class="anchor"></a>Constructing a matching</h2>
<p>Let’s consider two trees that differ in the position of one wildcard leaf, and in the resolution of a clade:</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">tree1</span> <span class="kw">&lt;-</span> <span class="kw pkg">ape</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span>(<span class="kw">text</span><span class="kw">=</span><span class="st">'((A, B), ((C, (D, E)), (F, (G, (H, I)))));'</span>)
<span class="no">tree2</span> <span class="kw">&lt;-</span> <span class="kw pkg">ape</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span>(<span class="kw">text</span><span class="kw">=</span><span class="st">'((A, B), ((C, D, (E, I)), (F, (G, H))));'</span>)

<span class="fu">Plot</span>(<span class="no">tree1</span>, <span class="no">tree2</span>, <span class="kw">highlight</span> <span class="kw">=</span> <span class="st">'I'</span>, <span class="kw">prune</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="fl">8</span>, <span class="fu"><a href="https://rdrr.io/r/base/integer.html">integer</a></span>(<span class="fl">0</span>)))</pre></body></html></div>
<p><img src="Robinson-Foulds_files/figure-html/unnamed-chunk-5-1.png" width="576" style="display: block; margin: auto;"></p>
<p>These trees obtain a Robinson-Foulds distance of nine: a large distances, as the maximum possible for trees of this resolution is eleven. <code>AB|CDEFGHI</code> is the only split in common between the two trees:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu">TwoTreePlot</span>()
<span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(<span class="no">RobinsonFouldsMatching</span>, <span class="no">tree1</span>, <span class="no">tree2</span>)</pre></body></html></div>
<p><img src="Robinson-Foulds_files/figure-html/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;"></p>
<p>This distance score is higher than might be expected, given how much the trees have in common; removing the single leaf ‘I’ results in two trees that differ only in the resolution of a single node:</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="fu">TwoTreePlot</span>()
<span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(<span class="no">RobinsonFouldsMatching</span>,
                  <span class="fu">drop.tip</span>(<span class="no">tree1</span>, <span class="st">'I'</span>),
                  <span class="fu">drop.tip</span>(<span class="no">tree2</span>, <span class="st">'I'</span>))</pre></body></html></div>
<p><img src="Robinson-Foulds_files/figure-html/unnamed-chunk-7-1.png" width="576" style="display: block; margin: auto;"></p>
<p>This hidden similarity can be better reflected if similar, but non-identical, splits are assigned non-zero similarity scores.</p>
<p>There are various ways to score the similarity between two splits. One is to build on the idea introduced above, where identical splits are scored according to their phylogenetic information content. Non-matching splits can be scored according to the amount of phylogenetic information that they hold in common, which is a function of the proportion of trees that are consistent with both splits. (A full explanation is provided in the discussion of <a href="Generalized-RF.html">Generalized Robinson-Foulds distances</a>.)</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu">TwoTreePlot</span>()
<span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(<span class="no">SharedPhylogeneticInfo</span>, <span class="no">tree1</span>, <span class="no">tree2</span>)</pre></body></html></div>
<p><img src="Robinson-Foulds_files/figure-html/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Here, the split <code>AB|CDEFGHI</code> occurs in both trees, and, as it happens, makes the largest contribution to the tree similarity score (3.70) for this particular pair of trees. This is the same contribution it would have made to the information-corrected Robinson-Foulds similarity.</p>
<p>The split <code>ABCDEF|GHI</code> in the left-hand tree is paired with the split <code>ABCDEFI|GH</code> in the right-hand tree. Had <code>ABCDEF|GHI</code> been available in the right-hand tree, then this perfect match would have been assigned a similarity of <code>SplitInformation(3, 6)</code> = 5.57 bits. The partial match is instead allocated a lower score of 2.12 bits. Pairings of incompatible splits, i.e. those that cannot co-exist on a tree, such as <code>ABCDEFG|HI</code> - <code>ABCDFGH|EI</code>, have no phylogenetic information in common. (<a href="information.html">clustering information</a> is an alternative way to think about split similarity that recognizes similarity even between incompatible splits.)</p>
<p>The matching depicted above is one of many. It happens to be optimal: an optimal matching can be found by considering the similarity score of each possible pairing, and solving a linear assignment problem to find the optimal set of pairings.</p>
<p>We can view the splits in each tree, named according to the number of their associated node:</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="kw pkg">TreeTools</span><span class="kw ns">::</span><span class="fu"><a href="https://ms609.github.io/TreeTools/reference/as.Splits.html">as.Splits</a></span>(<span class="no">tree1</span>, <span class="no">LETTERS</span>[<span class="fl">1</span>:<span class="fl">9</span>]))</pre></body></html></div>
<pre><code>## 6 bipartition splits dividing 9 tips, A .. I
##      123456789
##  11  **.......
##  13  ..***....
##  14  ...**....
##  15  .....****
##  16  ......***
##  17  .......**
## 
##  Tip 1: A     Tip 2: B    Tip 3: C    Tip 4: D    Tip 5: E   
##  Tip 6: F     Tip 7: G    Tip 8: H    Tip 9: I   </code></pre>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="kw pkg">TreeTools</span><span class="kw ns">::</span><span class="fu"><a href="https://ms609.github.io/TreeTools/reference/as.Splits.html">as.Splits</a></span>(<span class="no">tree2</span>, <span class="no">LETTERS</span>[<span class="fl">1</span>:<span class="fl">9</span>]))</pre></body></html></div>
<pre><code>## 5 bipartition splits dividing 9 tips, A .. I
##      123456789
##  11  **.......
##  13  ..***...*
##  14  ....*...*
##  15  .....***.
##  16  ......**.
## 
##  Tip 1: A     Tip 2: B    Tip 3: C    Tip 4: D    Tip 5: E   
##  Tip 6: F     Tip 7: G    Tip 8: H    Tip 9: I   </code></pre>
<p>We can then see the similarity scores for each pair of splits, along with the optimal matching:</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/attributes.html">attributes</a></span>(<span class="fu"><a href="../reference/TreeDistance.html">SharedPhylogeneticInfo</a></span>(<span class="no">tree1</span>, <span class="no">tree2</span>, <span class="kw">reportMatching</span> <span class="kw">=</span> <span class="fl">TRUE</span>))</pre></body></html></div>
<pre><code>## $matching
## [1]  1  2 NA  4  5  3
## 
## $pairScores
##           [,1]      [,2]      [,3]      [,4]      [,5]
## [1,] 3.7004397 0.8930848 0.2410081 0.5305147 0.2410081
## [2,] 0.5305147 3.2529807 0.0000000 1.1825914 0.5305147
## [3,] 0.2410081 1.3785116 0.0000000 0.5305147 0.2410081
## [4,] 0.8930848 0.0000000 0.0000000 3.2529807 1.3785116
## [5,] 0.5305147 0.0000000 0.0000000 0.0000000 2.1154772
## [6,] 0.2410081 0.0000000 0.0000000 0.0000000 0.0000000
## 
## $matchedSplits
## [1] "A B | C D E F G H I =&gt; A B | C D E F G H I"
## [2] "C D E | A B F G H I =&gt; C D E I | A B F G H"
## [3] "F G H I | A B C D E =&gt; F G H | A B C D E I"
## [4] "G H I | A B C D E F =&gt; G H | A B C D E F I"
## [5] "H I | A B C D E F G .. E I | A B C D F G H"</code></pre>
<p><code>..</code> denotes that the fifth matching contributes zero to similarity score; an alternative optimal matching would leave these splits unpaired.</p>
</div>
<div id="what-next" class="section level2">
<h2 class="hasAnchor">
<a href="#what-next" class="anchor"></a>What next?</h2>
<ul>
<li>Alternatives measures of split similarity, such as <a href="information.html">mutual clustering information</a>, give rise to other <a href="Generalized-RF.html">Generalized Robinson-Foulds distances</a>.</li>
</ul>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-Bocker2013">
<p>Böcker, S., Canzar, S., &amp; Klau, G. W. (2013). The generalized Robinson-Foulds metric. In A. Darling &amp; J. Stoye (Eds.), <em>Algorithms in Bioinformatics. WABI 2013. Lecture Notes in Computer Science, vol 8126</em> (pp. 156–169). Berlin, Heidelberg: Springer. doi:<a href="https://doi.org/10.1007/978-3-642-40453-5_13">10.1007/978-3-642-40453-5_13</a></p>
</div>
<div id="ref-Nye2006">
<p>Nye, T. M. W., Liò, P., &amp; Gilks, W. R. (2006). A novel algorithm and web-based tool for comparing two alternative phylogenetic trees. <em>Bioinformatics</em>, <em>22</em>(1), 117–119. doi:<a href="https://doi.org/10.1093/bioinformatics/bti720">10.1093/bioinformatics/bti720</a></p>
</div>
<div id="ref-Robinson1981">
<p>Robinson, D. F., &amp; Foulds, L. R. (1981). Comparison of phylogenetic trees. <em>Mathematical Biosciences</em>, <em>53</em>(1-2), 131–147. doi:<a href="https://doi.org/10.1016/0025-5564(81)90043-2">10.1016/0025-5564(81)90043-2</a></p>
</div>
<div id="ref-Schuh1980">
<p>Schuh, R. T., &amp; Polhemus, J. T. (1980). Analysis of taxonomic congruence among morphological, ecological and biogeographic data sets for the Leptopodomorpha (Hemiptera). <em>Systematic Biology</em>, <em>29</em>(1), 1–26. doi:<a href="https://doi.org/10.1093/sysbio/29.1.1">10.1093/sysbio/29.1.1</a></p>
</div>
<div id="ref-SmithDist">
<p>Smith, M. R. (2020). Information theoretic Generalized Robinson-Foulds metrics for comparing phylogenetic trees. <em>Bioinformatics</em>, online ahead of print. doi:<a href="https://doi.org/10.1093/bioinformatics/btaa614">10.1093/bioinformatics/btaa614</a></p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://community.dur.ac.uk/martin.smith/">Martin R. Smith</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '8bf44d312479af73720c5ca5fcc12eba',
    indexName: 'treedist',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
