<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Contextualizing tree distances • TreeDist</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Contextualizing tree distances">
<meta name="robots" content="noindex">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-BTKEG66D3F"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-BTKEG66D3F');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TreeDist</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">2.8.0.9002</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/Using-TreeDist.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Function reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tree distance analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/Using-TreeDist.html">Calculate tree similarity with 'TreeDist'</a></li>
    <li><a class="dropdown-item" href="../articles/using-distances.html">Contextualizing tree distances</a></li>
    <li><a class="dropdown-item" href="../articles/different-leaves.html">Trees with different leaves</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tree space analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/treespace.html">Tree space analysis</a></li>
    <li><a class="dropdown-item" href="../articles/landscapes.html">Analysing landscapes of phylogenetic trees</a></li>
    <li><a class="dropdown-item" href="../articles/compare-treesets.html">Comparing sets of trees from different analyses</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tree distance introductions</h6></li>
    <li><a class="dropdown-item" href="../articles/information.html">Comparing splits using information theory</a></li>
    <li><a class="dropdown-item" href="../articles/Robinson-Foulds.html">Extending the Robinson-Foulds metric</a></li>
    <li><a class="dropdown-item" href="../articles/Generalized-RF.html">Generalized Robinson-Foulds distances</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ms609/TreeDist"><span class="fa fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Contextualizing tree distances</h1>
                        <h4 data-toc-skip class="author"><a href="https://smithlabdurham.github.io/" class="external-link">Martin R. Smith</a></h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/ms609/TreeDist/blob/master/vignettes/using-distances.Rmd" class="external-link"><code>vignettes/using-distances.Rmd</code></a></small>
      <div class="d-none name"><code>using-distances.Rmd</code></div>
    </div>

    
    
<p>Once you understand <a href="Using-TreeDist.html">how to use
“TreeDist”</a> to calculate tree distances, the next step is to provide
some context for the calculated distances.</p>
<div class="section level2">
<h2 id="normalizing">Normalizing<a class="anchor" aria-label="anchor" href="#normalizing"></a>
</h2>
<p>The maximum value of most tree distance metrics scales with the size
of the trees being compared. Typically, the resolution of the trees also
impacts the range of possible values. As such, it can be difficult to
interpret the tree distance value without suitable context.</p>
<p>Normalizing a distance metric is one way to render its meaning more
obvious. Selecting an appropriate normalizing constant may require
careful consideration of the purpose to which a tree distance metric is
being put.</p>
<p>The default normalization behaviour of each function when
<code>normalize = TRUE</code> is listed in the <a href="../reference/index.html">function reference</a>, or can be viewed
by typing <code>?FunctionName</code> in the R terminal.</p>
<div class="section level3">
<h3 id="nye-et-al--tree-similarity">Nye <em>et al.</em> tree similarity<a class="anchor" aria-label="anchor" href="#nye-et-al--tree-similarity"></a>
</h3>
<p>Let’s work through a simple example using the Nye <em>et al</em>.
<span class="citation">(2006)</span> similarity metric to compare two
imperfectly-resolved trees.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ms609.github.io/TreeDist/">"TreeDist"</a></span><span class="op">)</span></span>
<span><span class="va">tree1</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">'(A, ((B, ((C, D), (E, F))), (G, (H, (I, J, K)))));'</span><span class="op">)</span></span>
<span><span class="va">tree2</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">'(A, (B, (C, D, E, (J, K)), (F, (G, H, I))));'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">NyeSimilarity</span>, <span class="va">tree1</span>, <span class="va">tree2</span>,</span>
<span>                  Plot <span class="op">=</span> <span class="va">TreeDistPlot</span>, matchZeros <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="using-distances_files/figure-html/unnamed-chunk-1-1.png" width="90%" style="display: block; margin: auto;"></p>
<p>This is a nice metric to start with, because the maximum similarity
between each pair of splits is defined as one. (Astute readers might
worry that the minimum similarity is greater than zero – that’s a harder
problem to overcome.)</p>
<p>As such, the maximum similarity possible between two 11-leaf trees is
<a href="https://ms609.github.io/TreeTools/reference/NSplits.html" class="external-link"><code>NSplits(11)</code></a>
= 8.</p>
<p>Normalizing against this value tells us how similar the two trees
are, compared to two identical eleven-leaf binary trees.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/NyeSimilarity.html">NyeSimilarity</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>, normalize <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">/</span> <span class="fl">8</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.2744048</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/NyeSimilarity.html">NyeSimilarity</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>, normalize <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.2744048</span></span></code></pre>
<p>This approach will result in a similarity score less than one if two
trees are identical, but not fully resolved (i.e. binary).<br>
As such, we might prefer to compare the tree similarity to the maximum
score possible for two trees of the specified resolution. This value is
given by the number of splits in the least resolved of the two
trees:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/NyeSimilarity.html">NyeSimilarity</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>,</span>
<span>                  normalize <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fu">TreeTools</span><span class="fu">::</span><span class="fu"><a href="https://ms609.github.io/TreeTools/reference/NSplits.html" class="external-link">NSplits</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.5488095</span></span></code></pre>
<p>More concisely, we can provide a normalizing function:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/NyeSimilarity.html">NyeSimilarity</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>, normalize <span class="op">=</span> <span class="va">min</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.2744048</span></span></code></pre>
<p>This approach will produce a similarity of one if one tree is a
less-resolved version of another (and thus not identical).</p>
<p>If we are comparing lists of trees, this best value will depend on
the number of splits in each pair of trees. We can use the function
<code><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmin()</a></code> to select the less resolved of each pair of
trees:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/NyeSimilarity.html">NyeSimilarity</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span><span class="op">)</span>, normalize <span class="op">=</span> <span class="va">pmin</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           [,1]      [,2]</span></span>
<span><span class="co">## [1,] 0.8750000 0.2744048</span></span>
<span><span class="co">## [2,] 0.2744048 0.5000000</span></span></code></pre>
<p>To avoid these limitations, we may instead opt to normalize against
the average number of splits in the two trees. This is the default
normalization method for <a href="../reference/NyeSimilarity.html"><code>NyeSimilarity()</code></a>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/NyeSimilarity.html">NyeSimilarity</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>, normalize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.3991342</span></span></code></pre>
<p>Finally, if <code>tree1</code> is a “target” tree – perhaps one that
has been used to simulate data from, or which is independently known to
be true or virtuous – we may wish to normalize against the best possible
match to that tree.</p>
<p>In that case, the best possible score is</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">TreeTools</span><span class="fu">::</span><span class="fu"><a href="https://ms609.github.io/TreeTools/reference/NSplits.html" class="external-link">NSplits</a></span><span class="op">(</span><span class="va">tree1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 7</span></span></code></pre>
<p>and our normalized score will be</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/NyeSimilarity.html">NyeSimilarity</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>, normalize <span class="op">=</span> <span class="fu">TreeTools</span><span class="fu">::</span><span class="fu"><a href="https://ms609.github.io/TreeTools/reference/NSplits.html" class="external-link">NSplits</a></span><span class="op">(</span><span class="va">tree1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.3136054</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="normalizing-to-random-similarity">Normalizing to random similarity<a class="anchor" aria-label="anchor" href="#normalizing-to-random-similarity"></a>
</h3>
<p>The diameter (maximum possible distance) of the Nye <em>et al</em>.
tree similarity metric is easy to calculate, but this is not the case
for all metrics. For example, the clustering information distance metric
<span class="citation">(Smith, 2020)</span> ranges in principle from
zero to the total clustering entropy present in a pair of trees. But
with even a modest number of leaves, no pairs of trees exist in which
every split in one tree is perfectly contradicted by every other split
in the other; as such, any pair of trees will necessarily have some
degree of similarity.<br>
In such a context, it can be relevant to normalize tree similarity
against the <em>expected</em> distance between a pair of random trees,
rather than a maximum value <span class="citation">(see Vinh, Epps,
&amp; Bailey, 2010)</span>. On this measure, distances greater than one
denote trees that are more different than expected by chance, whereas a
distance of zero denotes identity.</p>
<p>With the quartet divergence, the expected tree distance is readily
calculated: any given quartet has a one in three chance of matching by
chance.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ms609.github.io/Quartet/" class="external-link">"Quartet"</a></span>, exclude <span class="op">=</span> <span class="st">"RobinsonFoulds"</span><span class="op">)</span></span>
<span><span class="va">expectedQD</span> <span class="op">&lt;-</span> <span class="fl">2</span> <span class="op">/</span> <span class="fl">3</span></span>
<span><span class="va">normalizedQD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ms609.github.io/Quartet/reference/SimilarityMetrics.html" class="external-link">QuartetDivergence</a></span><span class="op">(</span><span class="fu"><a href="https://ms609.github.io/Quartet/reference/QuartetStatus.html" class="external-link">QuartetStatus</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span><span class="op">)</span>,</span>
<span>                                  similarity <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">/</span> <span class="va">expectedQD</span></span></code></pre></div>
<p>The expected distance is more difficult to calculate for other
metrics, but can be approximated by sampling random pairs of trees.
Measured distances between 10 000 pairs of random bifurcating trees with
up to 200 leaves are available in the data package ‘<a href="https://github.com/ms609/TreeDistData/" class="external-link">TreeDistData</a>’. We can
view (normalized) distances for a selection of methods:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"TreeDistData"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ms609.github.io/TreeDistData/">"TreeDistData"</a></span>, exclude <span class="op">=</span> <span class="st">"PairwiseDistances"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"randomTreeDistances"</span>, package <span class="op">=</span> <span class="st">"TreeDistData"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">methods</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pid"</span>, <span class="st">"cid"</span>, <span class="st">"nye"</span>, <span class="st">"qd"</span><span class="op">)</span></span>
<span>  <span class="va">methodCol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>pid <span class="op">=</span> <span class="st">"#e15659"</span>, cid <span class="op">=</span> <span class="st">"#58a14e"</span>, nye <span class="op">=</span> <span class="st">"#edc949"</span>,</span>
<span>                 qd <span class="op">=</span> <span class="st">"#af7aa1"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">oldPar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>cex <span class="op">=</span> <span class="fl">0.7</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">5</span>, <span class="fl">0.01</span>, <span class="fl">0.01</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">nLeaves</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dimnames.html" class="external-link">dimnames</a></span><span class="op">(</span><span class="va">randomTreeDistances</span><span class="op">)</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">nLeaves</span>, type <span class="op">=</span> <span class="st">"n"</span>, <span class="va">randomTreeDistances</span><span class="op">[</span><span class="st">"pid"</span>, <span class="st">"mean"</span>, <span class="op">]</span>,</span>
<span>       ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.54</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>       xlab <span class="op">=</span> <span class="st">"Number of leaves"</span>,</span>
<span>       ylab <span class="op">=</span> <span class="st">"Normalized distance between random tree pairs"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">method</span> <span class="kw">in</span> <span class="va">methods</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">dat</span> <span class="op">&lt;-</span> <span class="va">randomTreeDistances</span><span class="op">[</span><span class="va">method</span>, , <span class="op">]</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="va">nLeaves</span>, <span class="va">dat</span><span class="op">[</span><span class="st">"50%"</span>, <span class="op">]</span>, pch <span class="op">=</span> <span class="fl">1</span>, col <span class="op">=</span> <span class="va">methodCol</span><span class="op">[</span><span class="va">method</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/polygon.html" class="external-link">polygon</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">nLeaves</span>, <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">nLeaves</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">dat</span><span class="op">[</span><span class="st">"25%"</span>, <span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">dat</span><span class="op">[</span><span class="st">"75%"</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span>,</span>
<span>            border <span class="op">=</span> <span class="cn">NA</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">methodCol</span><span class="op">[</span><span class="va">method</span><span class="op">]</span>, <span class="st">"55"</span><span class="op">)</span><span class="op">)</span></span>
<span>    </span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/text.html" class="external-link">text</a></span><span class="op">(</span><span class="fl">202</span>, <span class="va">randomTreeDistances</span><span class="op">[</span><span class="va">methods</span>, <span class="st">"50%"</span>, <span class="st">"200"</span><span class="op">]</span> <span class="op">+</span> <span class="fl">0.02</span>, </span>
<span>       <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Different phylogenetic information"</span>, </span>
<span>         <span class="st">"Clustering information distance"</span>,</span>
<span>         <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu">plain</span><span class="op">(</span><span class="st">"Nye "</span><span class="op">)</span>, <span class="fu">italic</span><span class="op">(</span><span class="st">"et al."</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>         <span class="st">"Quartet divergence"</span></span>
<span>         <span class="op">)</span>, col <span class="op">=</span> <span class="va">methodCol</span><span class="op">[</span><span class="va">methods</span><span class="op">]</span>, pos <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span><span class="va">oldPar</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="using-distances_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
<p>or use these calculated values to normalize our tree distance:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expectedCID</span> <span class="op">&lt;-</span> <span class="va">randomTreeDistances</span><span class="op">[</span><span class="st">"cid"</span>, <span class="st">"mean"</span>, <span class="st">"9"</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/TreeDistance.html">ClusteringInfoDistance</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>, normalize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">/</span> <span class="va">expectedCID</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="testing-similarity-to-a-known-tree">Testing similarity to a known tree<a class="anchor" aria-label="anchor" href="#testing-similarity-to-a-known-tree"></a>
</h2>
<p>Similarity has two components: precision and accuracy <span class="citation">(Smith, 2019)</span>. A tree can be 80% similar to a
target tree because it contains 80% of the splits in the target tree,
and no incorrect splits – or because it is a binary tree in which 10% of
the splits present are resolved incorrectly and are thus positively
misleading.</p>
<p>In such a comparison, of course, it is more sensible to talk about
split <em>information</em> than just the number of splits: an even split
may contain more information than two very uneven splits, so the absence
of two information-poor splits may be preferable to the absence of one
information-rich split. As such, it is most instructive to think of the
proportion of information that has been correctly resolved: the goal is
to find a tree that is as informative as possible about the true
tree.</p>
<p>Ternary diagrams allow us to visualise the quality of a reconstructed
tree with reference to a known “true” tree:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">testTrees</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  trueTree <span class="op">=</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">'(a, (b, (c, (d, (e, (f, (g, h)))))));'</span><span class="op">)</span>,</span>
<span>  lackRes <span class="op">=</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">'(a, (b, c, (d, e, (f, g, h))));'</span><span class="op">)</span>,</span>
<span>  smallErr <span class="op">=</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">'(a, (c, (b, (d, (f, (e, (g, h)))))));'</span><span class="op">)</span>,</span>
<span>  bigErr <span class="op">=</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">'(a, (c, (((b, d), (f, h)), (e, g))));'</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">MutualClusteringInfo</span>, <span class="va">testTrees</span><span class="op">$</span><span class="va">trueTree</span>, <span class="va">testTrees</span><span class="op">$</span><span class="va">lackRes</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html" class="external-link">points</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">7.5</span>, pch <span class="op">=</span> <span class="fl">2</span>, cex <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="st">"#E69F00"</span>, xpd <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span></code></pre></div>
<p><img src="using-distances_files/figure-html/unnamed-chunk-12-1.png" width="80%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">MutualClusteringInfo</span>, <span class="va">testTrees</span><span class="op">$</span><span class="va">trueTree</span>, <span class="va">testTrees</span><span class="op">$</span><span class="va">smallErr</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html" class="external-link">points</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">7.5</span>, pch <span class="op">=</span> <span class="fl">3</span>, cex <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="st">"#56B4E9"</span>, xpd <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span></code></pre></div>
<p><img src="using-distances_files/figure-html/unnamed-chunk-12-2.png" width="80%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">MutualClusteringInfo</span>, <span class="va">testTrees</span><span class="op">$</span><span class="va">trueTree</span>, <span class="va">testTrees</span><span class="op">$</span><span class="va">bigErr</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html" class="external-link">points</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">7.5</span>, pch <span class="op">=</span> <span class="fl">4</span>, cex <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="st">"#009E73"</span>, xpd <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span></code></pre></div>
<p><img src="using-distances_files/figure-html/unnamed-chunk-12-3.png" width="80%" style="display: block; margin: auto;"></p>
<p>Better trees plot vertically towards the “100% shared information”
vertex. Resolution of trees increases towards the right; trees that are
more resolved may be no better than less-resolved trees if the addition
of resolution introduces error.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"Ternary"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ms609.github.io/Ternary/" class="external-link">"Ternary"</a></span><span class="op">)</span></span>
<span>  <span class="va">oldPar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/TernaryPlot.html" class="external-link">TernaryPlot</a></span><span class="op">(</span>alab <span class="op">=</span> <span class="st">"Absent information"</span>, blab <span class="op">=</span> <span class="st">"Shared information"</span>,</span>
<span>              clab <span class="op">=</span> <span class="st">"Misinformation"</span>,</span>
<span>              lab.cex <span class="op">=</span> <span class="fl">0.8</span>, lab.offset <span class="op">=</span> <span class="fl">0.18</span>,</span>
<span>              point <span class="op">=</span> <span class="st">"left"</span>, clockwise <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>              grid.col <span class="op">=</span> <span class="st">"#dedede"</span>, grid.minor.lines <span class="op">=</span> <span class="fl">0</span>,</span>
<span>              axis.labels <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">10</span> <span class="op">/</span> <span class="fl">10</span>, axis.col <span class="op">=</span> <span class="st">"#aaaaaa"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/TernaryPlot.html" class="external-link">HorizontalGrid</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="va">correct</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TreeDistance.html">MutualClusteringInfo</a></span><span class="op">(</span><span class="va">testTrees</span><span class="op">$</span><span class="va">trueTree</span>, <span class="va">testTrees</span><span class="op">)</span></span>
<span>  <span class="va">resolved</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TreeInfo.html">ClusteringEntropy</a></span><span class="op">(</span><span class="va">testTrees</span><span class="op">)</span></span>
<span>  <span class="va">unresolved</span> <span class="op">&lt;-</span> <span class="va">resolved</span><span class="op">[</span><span class="st">"trueTree"</span><span class="op">]</span> <span class="op">-</span> <span class="va">resolved</span></span>
<span>  <span class="va">incorrect</span> <span class="op">&lt;-</span> <span class="va">resolved</span> <span class="op">-</span> <span class="va">correct</span></span>
<span>  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/AddToTernary.html" class="external-link">TernaryPoints</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">unresolved</span>, <span class="va">correct</span>, <span class="va">incorrect</span><span class="op">)</span>, </span>
<span>                pch <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, cex <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="fu">Ternary</span><span class="fu">::</span><span class="va"><a href="https://ms609.github.io/Ternary/reference/cbPalettes.html" class="external-link">cbPalette8</a></span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span><span class="va">oldPar</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="using-distances_files/figure-html/unnamed-chunk-13-1.png" width="50%" style="display: block; margin: auto;"></p>
<div class="section level3">
<h3 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h3>
<p>Here’s a noddy real-world example applying this to a simulation-style
study.</p>
<p>First, let’s generate a starting tree, which will represent our
reference topology:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">trueTree</span> <span class="op">&lt;-</span> <span class="fu">TreeTools</span><span class="fu">::</span><span class="fu"><a href="https://ms609.github.io/TreeTools/reference/GenerateTree.html" class="external-link">RandomTree</a></span><span class="op">(</span><span class="fl">20</span>, root <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Then, let’s generate 200 degraded trees. We’ll move away from the
true tree by making a TBR move, then reduce resolution by taking the
consensus of this tree and three trees from its immediate neighbourhood
(one NNI move away).</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">treeSearchInstalled</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"TreeSearch"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">treeSearchInstalled</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ms609.github.io/TreeSearch/" class="external-link">"TreeSearch"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># for TBR, NNI</span></span>
<span>  <span class="va">oneAway</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/structure.html" class="external-link">structure</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">200</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">tbrTree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/TBR.html" class="external-link">TBR</a></span><span class="op">(</span><span class="va">trueTree</span><span class="op">)</span></span>
<span>    <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/consensus.html" class="external-link">consensus</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">tbrTree</span>,</span>
<span>                        <span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html" class="external-link">NNI</a></span><span class="op">(</span><span class="va">tbrTree</span><span class="op">)</span>,</span>
<span>                        <span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html" class="external-link">NNI</a></span><span class="op">(</span><span class="va">tbrTree</span><span class="op">)</span>,</span>
<span>                        <span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html" class="external-link">NNI</a></span><span class="op">(</span><span class="va">tbrTree</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span>, class <span class="op">=</span> <span class="st">"multiPhylo"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"Install \"TreeSearch\" to run this example"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>And let’s generate 200 more trees that are even more degraded. This
time we’ll move further (three TBR moves) from the true tree, and reduce
resolution by taking a consensus with three trees from its wider
neighbourhood (each two NNI moves away).</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">treeSearchInstalled</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">threeAway</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/structure.html" class="external-link">structure</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">200</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">tbrTree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/TBR.html" class="external-link">TBR</a></span><span class="op">(</span><span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/TBR.html" class="external-link">TBR</a></span><span class="op">(</span><span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/TBR.html" class="external-link">TBR</a></span><span class="op">(</span><span class="va">trueTree</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/consensus.html" class="external-link">consensus</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">tbrTree</span>, </span>
<span>                        <span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html" class="external-link">NNI</a></span><span class="op">(</span><span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html" class="external-link">NNI</a></span><span class="op">(</span><span class="va">tbrTree</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                        <span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html" class="external-link">NNI</a></span><span class="op">(</span><span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html" class="external-link">NNI</a></span><span class="op">(</span><span class="va">tbrTree</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                        <span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html" class="external-link">NNI</a></span><span class="op">(</span><span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html" class="external-link">NNI</a></span><span class="op">(</span><span class="va">tbrTree</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span>, class <span class="op">=</span> <span class="st">"multiPhylo"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Now let’s calculate their tree similarity scores. We need to
calculate the amount of information each tree has in common with the
true tree:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">treeSearchInstalled</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">correct1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TreeDistance.html">MutualClusteringInfo</a></span><span class="op">(</span><span class="va">trueTree</span>, <span class="va">oneAway</span><span class="op">)</span></span>
<span>  <span class="va">correct3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TreeDistance.html">MutualClusteringInfo</a></span><span class="op">(</span><span class="va">trueTree</span>, <span class="va">threeAway</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The amount of information in each degraded tree:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">treeSearchInstalled</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">infoInTree1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TreeInfo.html">ClusteringEntropy</a></span><span class="op">(</span><span class="va">oneAway</span><span class="op">)</span></span>
<span>  <span class="va">infoInTree3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TreeInfo.html">ClusteringEntropy</a></span><span class="op">(</span><span class="va">threeAway</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The amount of information that could have been resolved, but was
not:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">treeSearchInstalled</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">unresolved1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TreeInfo.html">ClusteringEntropy</a></span><span class="op">(</span><span class="va">trueTree</span><span class="op">)</span> <span class="op">-</span> <span class="va">infoInTree1</span></span>
<span>  <span class="va">unresolved3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TreeInfo.html">ClusteringEntropy</a></span><span class="op">(</span><span class="va">trueTree</span><span class="op">)</span> <span class="op">-</span> <span class="va">infoInTree3</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>And the amount of information incorrectly resolved:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">treeSearchInstalled</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">incorrect1</span> <span class="op">&lt;-</span> <span class="va">infoInTree1</span> <span class="op">-</span> <span class="va">correct1</span></span>
<span>  <span class="va">incorrect3</span> <span class="op">&lt;-</span> <span class="va">infoInTree3</span> <span class="op">-</span> <span class="va">correct3</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>In preparation for our plot, let’s colour our one-away trees <span style="background-color:#F5CCB599"> orange </span>, and our three-away
trees <span style="background-color:#9CE0E699"> blue </span>:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">col1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/hcl.html" class="external-link">hcl</a></span><span class="op">(</span><span class="fl">200</span>, alpha <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></span>
<span><span class="va">col3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/hcl.html" class="external-link">hcl</a></span><span class="op">(</span><span class="fl">40</span>, alpha <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></span>
<span><span class="va">spec1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/grDevices/col2rgb.html" class="external-link">col2rgb</a></span><span class="op">(</span><span class="va">col1</span>, alpha <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">4</span>, ncol <span class="op">=</span> <span class="fl">181</span><span class="op">)</span></span>
<span><span class="va">spec3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/grDevices/col2rgb.html" class="external-link">col2rgb</a></span><span class="op">(</span><span class="va">col3</span>, alpha <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">4</span>, ncol <span class="op">=</span> <span class="fl">181</span><span class="op">)</span></span>
<span><span class="va">spec1</span><span class="op">[</span><span class="fl">4</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="va">spec3</span><span class="op">[</span><span class="fl">4</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span><span class="op">:</span><span class="fl">180</span></span>
<span><span class="va">ColToHex</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/rgb.html" class="external-link">rgb</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="va">x</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>, <span class="va">x</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span>, maxColorValue <span class="op">=</span> <span class="fl">255</span><span class="op">)</span></span>
<span><span class="va">spec1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">spec1</span>, <span class="fl">2</span>, <span class="va">ColToHex</span><span class="op">)</span></span>
<span><span class="va">spec3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">spec3</span>, <span class="fl">2</span>, <span class="va">ColToHex</span><span class="op">)</span></span></code></pre></div>
<p>Now we can plot this information on a ternary diagram.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">treeSearchInstalled</span> <span class="op">&amp;&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"Ternary"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">layout</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>, widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">oldPar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/TernaryPlot.html" class="external-link">TernaryPlot</a></span><span class="op">(</span>alab <span class="op">=</span> <span class="st">"Information absent in degraded tree"</span>, </span>
<span>              blab <span class="op">=</span> <span class="st">"\n\nCorrect information in degraded tree"</span>, </span>
<span>              clab <span class="op">=</span> <span class="st">"Misinformation in degraded tree"</span>,</span>
<span>              point <span class="op">=</span> <span class="st">"left"</span>, clockwise <span class="op">=</span> <span class="cn">FALSE</span>, grid.minor.lines <span class="op">=</span> <span class="fl">0</span>,</span>
<span>              axis.labels <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">10</span> <span class="op">/</span> <span class="fl">10</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/TernaryPlot.html" class="external-link">HorizontalGrid</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">coords1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">unresolved1</span>, <span class="va">correct1</span>, <span class="va">incorrect1</span><span class="op">)</span></span>
<span>  <span class="va">coords3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">unresolved3</span>, <span class="va">correct3</span>, <span class="va">incorrect3</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/ColourTernary.html" class="external-link">ColourTernary</a></span><span class="op">(</span><span class="fu"><a href="https://ms609.github.io/Ternary/reference/TernaryPointValues.html" class="external-link">TernaryDensity</a></span><span class="op">(</span><span class="va">coords1</span>, resolution <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>, spectrum <span class="op">=</span> <span class="va">spec1</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/ColourTernary.html" class="external-link">ColourTernary</a></span><span class="op">(</span><span class="fu"><a href="https://ms609.github.io/Ternary/reference/TernaryPointValues.html" class="external-link">TernaryDensity</a></span><span class="op">(</span><span class="va">coords3</span>, resolution <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>, spectrum <span class="op">=</span> <span class="va">spec3</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/TernaryDensityContour.html" class="external-link">TernaryDensityContour</a></span><span class="op">(</span><span class="va">coords3</span>, col <span class="op">=</span> <span class="va">col3</span>, nlevels <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/TernaryDensityContour.html" class="external-link">TernaryDensityContour</a></span><span class="op">(</span><span class="va">coords1</span>, col <span class="op">=</span> <span class="va">col1</span>, nlevels <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"kdensity"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/JonasMoss/kdensity" class="external-link">"kdensity"</a></span><span class="op">)</span></span>
<span>    <span class="va">HorizontalKDE</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">dat</span>, <span class="va">col</span>, <span class="va">add</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="va">lty</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span>      <span class="va">lwd</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span>      <span class="va">kde</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/kdensity/man/kdensity.html" class="external-link">kdensity</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></span>
<span>      <span class="va">kdeRange</span> <span class="op">&lt;-</span> <span class="fu">kdensity</span><span class="fu">:::</span><span class="fu"><a href="https://rdrr.io/pkg/kdensity/man/get_range.html" class="external-link">get_range</a></span><span class="op">(</span><span class="va">kde</span><span class="op">)</span></span>
<span>      <span class="kw">if</span> <span class="op">(</span><span class="va">add</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="fu">kde</span><span class="op">(</span><span class="va">kdeRange</span><span class="op">)</span>, <span class="va">kdeRange</span>, col <span class="op">=</span> <span class="va">col</span>, lty <span class="op">=</span> <span class="va">lty</span>, lwd <span class="op">=</span> <span class="va">lwd</span><span class="op">)</span></span>
<span>      <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu">kde</span><span class="op">(</span><span class="va">kdeRange</span><span class="op">)</span>, <span class="va">kdeRange</span>, col <span class="op">=</span> <span class="va">col</span>, lty <span class="op">=</span> <span class="va">lty</span>, lwd <span class="op">=</span> <span class="va">lwd</span>, </span>
<span>             ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">""</span>, axes <span class="op">=</span> <span class="cn">FALSE</span>, type <span class="op">=</span> <span class="st">"l"</span><span class="op">)</span></span>
<span>      <span class="op">}</span></span>
<span>      <span class="co"># abline(h = 0:10 / 10) # Useful for confirming alignment</span></span>
<span>    <span class="op">}</span></span>
<span>  </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.8</span>, <span class="fl">0</span>, <span class="fl">1.8</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="co"># align plot limits with ternary plot</span></span>
<span>    <span class="fu">HorizontalKDE</span><span class="op">(</span><span class="va">correct1</span> <span class="op">/</span> <span class="va">infoInTree1</span>, <span class="va">col1</span>, add <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    <span class="fu">HorizontalKDE</span><span class="op">(</span><span class="va">correct3</span> <span class="op">/</span> <span class="va">infoInTree3</span>, <span class="va">col3</span>, add <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/mtext.html" class="external-link">mtext</a></span><span class="op">(</span><span class="st">"\u2192 Normalized tree quality \u2192"</span>, <span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span><span class="va">oldPar</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"Install \"TreeSearch\" and \"Ternary\" to generate this plot"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="using-distances_files/figure-html/unnamed-chunk-22-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>In the ternary plot, the vertical direction corresponds to the
normalized tree quality, as depicted in the accompanying histogram.</p>
</div>
</div>
<div class="section level2">
<h2 id="what-next">What next?<a class="anchor" aria-label="anchor" href="#what-next"></a>
</h2>
<p>You may wish to:</p>
<ul>
<li><p>Explore the <a href="https://ms609.github.io/Ternary/" class="external-link">Ternary
package</a></p></li>
<li><p><a href="https://ms609.github.io/TreeDistData/articles/09-expected-similarity.html">Interpret
tree distance metrics</a></p></li>
<li><p>Compare trees with <a href="different-leaves.html">different
tips</a></p></li>
<li><p>Review <a href="https://ms609.github.io/TreeDist/index.html">available distance
measures</a> and the corresponding <a href="https://ms609.github.io/TreeDist/reference/index.html#section-tree-distance-measures">TreeDist
functions</a></p></li>
<li><p>Construct <a href="treespace.html">tree spaces</a> to visualize
landscapes of phylogenetic trees</p></li>
</ul>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0" line-spacing="2">
<div id="ref-Nye2006" class="csl-entry">
Nye, T. M. W., Liò, P., &amp; Gilks, W. R. (2006). <span class="nocase">A novel algorithm and web-based tool for comparing two
alternative phylogenetic trees</span>. <em>Bioinformatics</em>,
<em>22</em>(1), 117–119. doi:<a href="https://doi.org/10.1093/bioinformatics/bti720" class="external-link">10.1093/bioinformatics/bti720</a>
</div>
<div id="ref-Smith2019" class="csl-entry">
Smith, M. R. (2019). Bayesian and parsimony approaches reconstruct
informative trees from simulated morphological datasets. <em>Biology
Letters</em>, <em>15</em>(2), 20180632. doi:<a href="https://doi.org/10.1098/rsbl.2018.0632" class="external-link">10.1098/rsbl.2018.0632</a>
</div>
<div id="ref-SmithDist" class="csl-entry">
Smith, M. R. (2020). Information theoretic <span>Generalized
Robinson-Foulds</span> metrics for comparing phylogenetic trees.
<em>Bioinformatics</em>, <em>36</em>(20), 5007–5013. doi:<a href="https://doi.org/10.1093/bioinformatics/btaa614" class="external-link">10.1093/bioinformatics/btaa614</a>
</div>
<div id="ref-Vinh2010" class="csl-entry">
Vinh, N. X., Epps, J., &amp; Bailey, J. (2010). <span class="nocase">Information theoretic measures for clusterings
comparison: variants, properties, normalization and correction for
chance</span>. <em>Journal of Machine Learning Research</em>,
<em>11</em>, 2837–2854. doi:<a href="https://doi.org/10.1145/1553374.1553511" class="external-link">10.1145/1553374.1553511</a>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://smithlabdurham.github.io/" class="external-link">Martin R. Smith</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
