<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Contextualizing tree distances • TreeDist</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Contextualizing tree distances">
<meta property="og:description" content="TreeDist">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-156666971-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156666971-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TreeDist</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="In-development version">1.2.1.9201</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/Using-TreeDist.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Using-TreeDist.html">Using TreeDist to calculate tree similarity</a>
    </li>
    <li>
      <a href="../articles/treespace.html">Tree space analysis in TreeDist</a>
    </li>
    <li>
      <a href="../articles/using-distances.html">Contextualizing tree distances</a>
    </li>
    <li>
      <a href="../articles/Robinson-Foulds.html">Robinson-Foulds metric</a>
    </li>
    <li>
      <a href="../articles/Generalized-RF.html">Generalized Robinson-Foulds distances</a>
    </li>
    <li>
      <a href="../articles/information.html">Concepts of information</a>
    </li>
    <li>
      <a href="../articles/Tree-distance-interpretation.html">Evaluating tree distance metrics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ms609/TreeDist">
    <span class="fas fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="using-distances_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Contextualizing tree distances</h1>
                        <h4 class="author"><a href="https://smithlabdurham.github.io/">Martin R. Smith</a></h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/ms609/TreeDist/blob/master/vignettes/using-distances.Rmd"><code>vignettes/using-distances.Rmd</code></a></small>
      <div class="hidden name"><code>using-distances.Rmd</code></div>

    </div>

    
    
<p>Once you understand <a href="Using-TreeDist.html">how to use ‘TreeDist’</a> to calculate tree distances, the next step is to provide some context for the calculated distances.</p>
<div id="normalizing" class="section level2">
<h2 class="hasAnchor">
<a href="#normalizing" class="anchor"></a>Normalizing</h2>
<p>The maximum value of most tree distance metrics scales with the size of the trees being compared. Typically, the resolution of the trees also impacts the range of possible values. As such, it can be difficult to interpret the tree distance value without suitable context.</p>
<p>Normalizing a distance metric is one way to render its meaning more obvious. Selecting an appropriate normalizing constant may require careful consideration of the purpose to which a tree distance metric is being put.</p>
<p>The default normalization behaviour of each function when <code>normalize = TRUE</code> is listed in the <a href="../reference/index.html">function reference</a>, or can be viewed by typing <code>?FunctionName</code> in the R terminal.</p>
<div id="nye-et-al--tree-similarity" class="section level3">
<h3 class="hasAnchor">
<a href="#nye-et-al--tree-similarity" class="anchor"></a>Nye <em>et al.</em> tree similarity</h3>
<p>Let’s work through a simple example using the Nye <em>et al</em>. <span class="citation">(2006)</span> similarity metric to compare two imperfectly-resolved trees.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://ms609.github.io/TreeDist/">'TreeDist'</a></span><span class="op">)</span>
<span class="va">tree1</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">'(A, ((B, ((C, D), (E, F))), (G, (H, (I, J, K)))));'</span><span class="op">)</span>
<span class="va">tree2</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">'(A, (B, (C, D, E, (J, K)), (F, (G, H, I))));'</span><span class="op">)</span>
<span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">NyeSimilarity</span>, <span class="va">tree1</span>, <span class="va">tree2</span>,
                  Plot <span class="op">=</span> <span class="va">TreeDistPlot</span>, matchZeros <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p><img src="using-distances_files/figure-html/unnamed-chunk-1-1.png" width="90%" style="display: block; margin: auto;"></p>
<p>This is a nice metric to start with, because the maximum similarity between each pair of splits is defined as one. (Astute readers might worry that the minimum similarity is greater than zero – that’s a harder problem to overcome.)</p>
<p>As such, the maximum similarity possible between two 11-leaf trees is <a href="https://ms609.github.io/TreeTools/reference/NSplits.html"><code>NSplits(11)</code></a> = 8.</p>
<p>Normalizing against this value tells us how similar the two trees are, compared to two identical eleven-leaf binary trees.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/NyeSimilarity.html">NyeSimilarity</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>, normalize <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">/</span> <span class="fl">8</span></code></pre></div>
<pre><code>## [1] 0.2744048</code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/NyeSimilarity.html">NyeSimilarity</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>, normalize <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 0.2744048</code></pre>
<p>This approach will result in a similarity score less than one if two trees are identical, but not fully resolved (i.e. binary).<br>
As such, we might prefer to compare the tree similarity to the maximum score possible for two trees of the specified resolution. This value is given by the number of splits in the least resolved of the two trees:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/NyeSimilarity.html">NyeSimilarity</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>,
                  normalize <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="fu">TreeTools</span><span class="fu">::</span><span class="fu"><a href="https://ms609.github.io/TreeTools/reference/NSplits.html">NSplits</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 0.5488095</code></pre>
<p>More concisely, we can provide a normalizing function:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/NyeSimilarity.html">NyeSimilarity</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>, normalize <span class="op">=</span> <span class="va">min</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 0.2744048</code></pre>
<p>This approach will produce a similarity of one if one tree is a less-resolved version of another (and thus not identical).</p>
<p>If we are comparing lists of trees, this best value will depend on the number of splits in each pair of trees. We can use the function <code><a href="https://rdrr.io/r/base/Extremes.html">pmin()</a></code> to select the less resolved of each pair of trees:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/NyeSimilarity.html">NyeSimilarity</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span><span class="op">)</span>, normalize <span class="op">=</span> <span class="va">pmin</span><span class="op">)</span></code></pre></div>
<pre><code>##           [,1]      [,2]
## [1,] 0.8750000 0.2744048
## [2,] 0.2744048 0.5000000</code></pre>
<p>To avoid these limitations, we may instead opt to normalize against the average number of splits in the two trees. This is the default normalization method for <a href="../reference/NyeSimilarity.html"><code>NyeSimilarity()</code></a>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/NyeSimilarity.html">NyeSimilarity</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>, normalize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 0.3991342</code></pre>
<p>Finally, if <code>tree1</code> is a ‘target’ tree – perhaps one that has been used to simulate data from, or which is independently known to be true or virtuous – we may wish to normalize against the best possible match to that tree.</p>
<p>In that case, the best possible score is</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">TreeTools</span><span class="fu">::</span><span class="fu"><a href="https://ms609.github.io/TreeTools/reference/NSplits.html">NSplits</a></span><span class="op">(</span><span class="va">tree1</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 7</code></pre>
<p>and our normalized score will be</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/NyeSimilarity.html">NyeSimilarity</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>, normalize <span class="op">=</span> <span class="fu">TreeTools</span><span class="fu">::</span><span class="fu"><a href="https://ms609.github.io/TreeTools/reference/NSplits.html">NSplits</a></span><span class="op">(</span><span class="va">tree1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 0.3136054</code></pre>
</div>
<div id="normalizing-to-random-similarity" class="section level3">
<h3 class="hasAnchor">
<a href="#normalizing-to-random-similarity" class="anchor"></a>Normalizing to random similarity</h3>
<p>The diameter (maximum possible distance) of the Nye <em>et al</em>. tree similarity metric is easy to calculate, but this is not the case for all metrics. For example, the clustering information distance metric <span class="citation">(Smith, 2020)</span> ranges in principle from zero to the total clustering entropy present in a pair of trees. But with even a modest number of leaves, no pairs of trees exist in which every split in one tree is perfectly contradicted by every other split in the other; as such, any pair of trees will necessarily have some degree of similarity.<br>
In such a context, it can be relevant to normalize tree similarity against the <em>expected</em> distance between a pair of random trees, rather than a maximum value <span class="citation">(see Vinh, Epps, &amp; Bailey, 2010)</span>. On this measure, distances greater than one denote trees that are more different than expected by chance, whereas a distance of zero denotes identity.</p>
<p>With the quartet divergence, the expected tree distance is readily calculated: any given quartet has a one in three chance of matching by chance.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://ms609.github.io/Quartet/">'Quartet'</a></span>, exclude <span class="op">=</span> <span class="st">'RobinsonFoulds'</span><span class="op">)</span>
<span class="va">expectedQD</span> <span class="op">&lt;-</span> <span class="fl">2</span> <span class="op">/</span> <span class="fl">3</span>
<span class="va">normalizedQD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ms609.github.io/Quartet/reference/SimilarityMetrics.html">QuartetDivergence</a></span><span class="op">(</span><span class="fu"><a href="https://ms609.github.io/Quartet/reference/QuartetStatus.html">QuartetStatus</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span><span class="op">)</span>,
                                  similarity <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">/</span> <span class="va">expectedQD</span></code></pre></div>
<p>The expected distance is more difficult to calculate for other metrics, but can be approximated by sampling random pairs of trees. Measured distances between 10 000 pairs of random bifurcating trees with up to 200 leaves are available in the data package ‘<a href="https://github.com/ms609/TreeDistData/">TreeDistData</a>’. We can view (normalized) distances for a selection of methods:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">'TreeDistData'</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://ms609.github.io/TreeDistData/">'TreeDistData'</a></span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">'randomTreeDistances'</span>, package <span class="op">=</span> <span class="st">'TreeDistData'</span><span class="op">)</span>
  
  <span class="va">methods</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'pid'</span>, <span class="st">'cid'</span>, <span class="st">'nye'</span>, <span class="st">'qd'</span><span class="op">)</span>
  <span class="va">methodCol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>pid <span class="op">=</span> <span class="st">'#e15659'</span>, cid <span class="op">=</span> <span class="st">'#58a14e'</span>, nye <span class="op">=</span> <span class="st">'#edc949'</span>,
                 qd <span class="op">=</span> <span class="st">'#af7aa1'</span><span class="op">)</span>
  
  <span class="va">oldPar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>cex <span class="op">=</span> <span class="fl">0.7</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">5</span>, <span class="fl">0.01</span>, <span class="fl">0.01</span><span class="op">)</span><span class="op">)</span>
  <span class="va">nLeaves</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dimnames.html">dimnames</a></span><span class="op">(</span><span class="va">randomTreeDistances</span><span class="op">)</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">nLeaves</span>, type <span class="op">=</span> <span class="st">'n'</span>, <span class="va">randomTreeDistances</span><span class="op">[</span><span class="st">'pid'</span>, <span class="st">'mean'</span>, <span class="op">]</span>,
       ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.54</span>, <span class="fl">1</span><span class="op">)</span>,
       xlab <span class="op">=</span> <span class="st">"Number of leaves"</span>,
       ylab <span class="op">=</span> <span class="st">"Normalized distance between random tree pairs"</span><span class="op">)</span>
  
  <span class="kw">for</span> <span class="op">(</span><span class="va">method</span> <span class="kw">in</span> <span class="va">methods</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">dat</span> <span class="op">&lt;-</span> <span class="va">randomTreeDistances</span><span class="op">[</span><span class="va">method</span>, , <span class="op">]</span>
    <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="va">nLeaves</span>, <span class="va">dat</span><span class="op">[</span><span class="st">'50%'</span>, <span class="op">]</span>, pch <span class="op">=</span> <span class="fl">1</span>, col <span class="op">=</span> <span class="va">methodCol</span><span class="op">[</span><span class="va">method</span><span class="op">]</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/graphics/polygon.html">polygon</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">nLeaves</span>, <span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span><span class="op">(</span><span class="va">nLeaves</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">dat</span><span class="op">[</span><span class="st">'25%'</span>, <span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span><span class="op">(</span><span class="va">dat</span><span class="op">[</span><span class="st">'75%'</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span>,
            border <span class="op">=</span> <span class="cn">NA</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">methodCol</span><span class="op">[</span><span class="va">method</span><span class="op">]</span>, <span class="st">'55'</span><span class="op">)</span><span class="op">)</span>
    
  <span class="op">}</span>
  
  <span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span><span class="op">(</span><span class="fl">202</span>, <span class="va">randomTreeDistances</span><span class="op">[</span><span class="va">methods</span>, <span class="st">'50%'</span>, <span class="st">'200'</span><span class="op">]</span> <span class="op">+</span> <span class="fl">0.02</span>, 
       <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Different phylogenetic information'</span>, 
         <span class="st">'Clustering information distance'</span>,
         <span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/grDevices/plotmath.html">plain</a></span><span class="op">(</span><span class="st">'Nye '</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/grDevices/plotmath.html">italic</a></span><span class="op">(</span><span class="st">'et al.'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,
         <span class="st">'Quartet divergence'</span>
         <span class="op">)</span>, col <span class="op">=</span> <span class="va">methodCol</span><span class="op">[</span><span class="va">methods</span><span class="op">]</span>, pos <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span><span class="va">oldPar</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="using-distances_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
<p>or use these calculated values to normalize our tree distance:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">expectedCID</span> <span class="op">&lt;-</span> <span class="va">randomTreeDistances</span><span class="op">[</span><span class="st">'cid'</span>, <span class="st">'mean'</span>, <span class="st">'9'</span><span class="op">]</span>
<span class="fu"><a href="../reference/TreeDistance.html">ClusteringInfoDistance</a></span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>, normalize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">/</span> <span class="va">expectedCID</span></code></pre></div>
</div>
</div>
<div id="testing-similarity-to-a-known-tree" class="section level2">
<h2 class="hasAnchor">
<a href="#testing-similarity-to-a-known-tree" class="anchor"></a>Testing similarity to a known tree</h2>
<p>Similarity has two components: precision and accuracy <span class="citation">(Smith, 2019)</span>. A tree can be 80% similar to a target tree because it contains 80% of the splits in the target tree, and no incorrect splits – or because it is a binary tree in which 10% of the splits present are resolved incorrectly and are thus positively misleading.</p>
<p>In such a comparison, of course, it is more sensible to talk about split <em>information</em> than just the number of splits: an even split may contain more information than two very uneven splits, so the absence of two information-poor splits may be preferable to the absence of one information-rich split. As such, it is most instructive to think of the proportion of information that has been correctly resolved: the goal is to find a tree that is as informative as possible about the true tree.</p>
<p>Ternary diagrams allow us to visualise the quality of a reconstructed tree with reference to a known ‘true’ tree:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">testTrees</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
  trueTree <span class="op">=</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">'(a, (b, (c, (d, (e, (f, (g, h)))))));'</span><span class="op">)</span>,
  lackRes <span class="op">=</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">'(a, (b, c, (d, e, (f, g, h))));'</span><span class="op">)</span>,
  smallErr <span class="op">=</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">'(a, (c, (b, (d, (f, (e, (g, h)))))));'</span><span class="op">)</span>,
  bigErr <span class="op">=</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">'(a, (c, (((b, d), (f, h)), (e, g))));'</span><span class="op">)</span>
<span class="op">)</span>
<span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">MutualClusteringInfo</span>, <span class="va">testTrees</span><span class="op">$</span><span class="va">trueTree</span>, <span class="va">testTrees</span><span class="op">$</span><span class="va">lackRes</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">7.5</span>, pch <span class="op">=</span> <span class="fl">2</span>, cex <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="st">"#E69F00"</span><span class="op">)</span></code></pre></div>
<p><img src="using-distances_files/figure-html/unnamed-chunk-12-1.png" width="80%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">MutualClusteringInfo</span>, <span class="va">testTrees</span><span class="op">$</span><span class="va">trueTree</span>, <span class="va">testTrees</span><span class="op">$</span><span class="va">smallErr</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">7.5</span>, pch <span class="op">=</span> <span class="fl">3</span>, cex <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="st">"#56B4E9"</span><span class="op">)</span></code></pre></div>
<p><img src="using-distances_files/figure-html/unnamed-chunk-12-2.png" width="80%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">MutualClusteringInfo</span>, <span class="va">testTrees</span><span class="op">$</span><span class="va">trueTree</span>, <span class="va">testTrees</span><span class="op">$</span><span class="va">bigErr</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">7.5</span>, pch <span class="op">=</span> <span class="fl">4</span>, cex <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="st">"#009E73"</span><span class="op">)</span></code></pre></div>
<p><img src="using-distances_files/figure-html/unnamed-chunk-12-3.png" width="80%" style="display: block; margin: auto;"></p>
<p>Better trees plot vertically towards the ‘100% shared information’ vertex. Resolution of trees increases towards the right; trees that are more resolved may be no better than less-resolved trees if the addition of resolution introduces error.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">'Ternary'</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://ms609.github.io/Ternary/">'Ternary'</a></span><span class="op">)</span>
  <span class="va">oldPar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/TernaryPlot.html">TernaryPlot</a></span><span class="op">(</span>alab <span class="op">=</span> <span class="st">'Absent information'</span>, blab <span class="op">=</span> <span class="st">'Shared information'</span>,
              clab <span class="op">=</span> <span class="st">'Misinformation'</span>,
              lab.cex <span class="op">=</span> <span class="fl">0.8</span>, lab.offset <span class="op">=</span> <span class="fl">0.18</span>,
              point <span class="op">=</span> <span class="st">'left'</span>, clockwise <span class="op">=</span> <span class="cn">FALSE</span>,
              grid.col <span class="op">=</span> <span class="st">'#dedede'</span>, grid.minor.lines <span class="op">=</span> <span class="fl">0</span>,
              axis.labels <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">10</span> <span class="op">/</span> <span class="fl">10</span>, axis.col <span class="op">=</span> <span class="st">"#aaaaaa"</span><span class="op">)</span>
  
  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/TernaryPlot.html">HorizontalGrid</a></span><span class="op">(</span><span class="op">)</span>
  <span class="va">correct</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TreeDistance.html">MutualClusteringInfo</a></span><span class="op">(</span><span class="va">testTrees</span><span class="op">$</span><span class="va">trueTree</span>, <span class="va">testTrees</span><span class="op">)</span>
  <span class="va">resolved</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ClusteringEntropy.html">ClusteringEntropy</a></span><span class="op">(</span><span class="va">testTrees</span><span class="op">)</span>
  <span class="va">unresolved</span> <span class="op">&lt;-</span> <span class="va">resolved</span><span class="op">[</span><span class="st">'trueTree'</span><span class="op">]</span> <span class="op">-</span> <span class="va">resolved</span>
  <span class="va">incorrect</span> <span class="op">&lt;-</span> <span class="va">resolved</span> <span class="op">-</span> <span class="va">correct</span>
  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/AddToTernary.html">TernaryPoints</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">unresolved</span>, <span class="va">correct</span>, <span class="va">incorrect</span><span class="op">)</span>, 
                pch <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, cex <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="fu">Ternary</span><span class="fu">::</span><span class="va"><a href="https://ms609.github.io/Ternary/reference/cbPalettes.html">cbPalette8</a></span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span><span class="va">oldPar</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="using-distances_files/figure-html/unnamed-chunk-13-1.png" width="50%" style="display: block; margin: auto;"></p>
<div id="example" class="section level3">
<h3 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h3>
<p>Here’s a noddy real-world example applying this to a simulation-style study.</p>
<p>First, let’s generate a starting tree, which will represent our reference topology:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>
<span class="va">trueTree</span> <span class="op">&lt;-</span> <span class="fu">TreeTools</span><span class="fu">::</span><span class="fu"><a href="https://ms609.github.io/TreeTools/reference/GenerateTree.html">RandomTree</a></span><span class="op">(</span><span class="fl">20</span>, root <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Then, let’s generate 200 degraded trees. We’ll move away from the true tree by making a TBR move, then reduce resolution by taking the consensus of this tree and three trees from its immediate neighbourhood (one NNI move away).</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://ms609.github.io/TreeSearch">'TreeSearch'</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span>, warn.conflict <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># for TBR, NNI</span>
<span class="va">oneAway</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/structure.html">structure</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span><span class="op">(</span><span class="fl">200</span><span class="op">)</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">tbrTree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/TBR.html">TBR</a></span><span class="op">(</span><span class="va">trueTree</span><span class="op">)</span>
  <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/consensus.html">consensus</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">tbrTree</span>,
                      <span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html">NNI</a></span><span class="op">(</span><span class="va">tbrTree</span><span class="op">)</span>,
                      <span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html">NNI</a></span><span class="op">(</span><span class="va">tbrTree</span><span class="op">)</span>,
                      <span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html">NNI</a></span><span class="op">(</span><span class="va">tbrTree</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>, class <span class="op">=</span> <span class="st">'multiPhylo'</span><span class="op">)</span></code></pre></div>
<p>And let’s generate 200 more trees that are even more degraded. This time we’ll move further (three TBR moves) from the true tree, and reduce resolution by taking a consensus with three trees from its wider neighbourhood (each two NNI moves away).</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">threeAway</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/structure.html">structure</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span><span class="op">(</span><span class="fl">200</span><span class="op">)</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">tbrTree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/TBR.html">TBR</a></span><span class="op">(</span><span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/TBR.html">TBR</a></span><span class="op">(</span><span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/TBR.html">TBR</a></span><span class="op">(</span><span class="va">trueTree</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
  <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/consensus.html">consensus</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">tbrTree</span>, 
                      <span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html">NNI</a></span><span class="op">(</span><span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html">NNI</a></span><span class="op">(</span><span class="va">tbrTree</span><span class="op">)</span><span class="op">)</span>,
                      <span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html">NNI</a></span><span class="op">(</span><span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html">NNI</a></span><span class="op">(</span><span class="va">tbrTree</span><span class="op">)</span><span class="op">)</span>,
                      <span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html">NNI</a></span><span class="op">(</span><span class="fu"><a href="https://ms609.github.io/TreeSearch/reference/NNI.html">NNI</a></span><span class="op">(</span><span class="va">tbrTree</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>, class <span class="op">=</span> <span class="st">'multiPhylo'</span><span class="op">)</span></code></pre></div>
<p>Now let’s calculate their tree similarity scores. We need to calculate the amount of information each tree has in common with the true tree:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">correct1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TreeDistance.html">MutualClusteringInfo</a></span><span class="op">(</span><span class="va">trueTree</span>, <span class="va">oneAway</span><span class="op">)</span>
<span class="va">correct3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TreeDistance.html">MutualClusteringInfo</a></span><span class="op">(</span><span class="va">trueTree</span>, <span class="va">threeAway</span><span class="op">)</span></code></pre></div>
<p>The amount of information in each degraded tree:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">infoInTree1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ClusteringEntropy.html">ClusteringEntropy</a></span><span class="op">(</span><span class="va">oneAway</span><span class="op">)</span>
<span class="va">infoInTree3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ClusteringEntropy.html">ClusteringEntropy</a></span><span class="op">(</span><span class="va">threeAway</span><span class="op">)</span></code></pre></div>
<p>The amount of information that could have been resolved, but was not:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">unresolved1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ClusteringEntropy.html">ClusteringEntropy</a></span><span class="op">(</span><span class="va">trueTree</span><span class="op">)</span> <span class="op">-</span> <span class="va">infoInTree1</span>
<span class="va">unresolved3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ClusteringEntropy.html">ClusteringEntropy</a></span><span class="op">(</span><span class="va">trueTree</span><span class="op">)</span> <span class="op">-</span> <span class="va">infoInTree3</span></code></pre></div>
<p>And the amount of information incorrectly resolved:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">incorrect1</span> <span class="op">&lt;-</span> <span class="va">infoInTree1</span> <span class="op">-</span> <span class="va">correct1</span>
<span class="va">incorrect3</span> <span class="op">&lt;-</span> <span class="va">infoInTree3</span> <span class="op">-</span> <span class="va">correct3</span></code></pre></div>
<p>In preparation for our plot, let’s colour our one-away trees <span style="background-color:#F5CCB599"> orange </span>, and our three-away trees <span style="background-color:#9CE0E699"> blue </span>:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">col1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/hcl.html">hcl</a></span><span class="op">(</span><span class="fl">200</span>, alpha <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span>
<span class="va">col3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/hcl.html">hcl</a></span><span class="op">(</span><span class="fl">40</span>, alpha <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span>
<span class="va">spec1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/grDevices/col2rgb.html">col2rgb</a></span><span class="op">(</span><span class="va">col1</span>, alpha <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">4</span>, ncol <span class="op">=</span> <span class="fl">181</span><span class="op">)</span>
<span class="va">spec3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/grDevices/col2rgb.html">col2rgb</a></span><span class="op">(</span><span class="va">col3</span>, alpha <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">4</span>, ncol <span class="op">=</span> <span class="fl">181</span><span class="op">)</span>
<span class="va">spec1</span><span class="op">[</span><span class="fl">4</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="va">spec3</span><span class="op">[</span><span class="fl">4</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span><span class="op">:</span><span class="fl">180</span>
<span class="va">ColToHex</span> <span class="op">&lt;-</span> <span class="kw">function</span> <span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/rgb.html">rgb</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="va">x</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>, <span class="va">x</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span>, maxColorValue <span class="op">=</span> <span class="fl">255</span><span class="op">)</span>
<span class="va">spec1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">spec1</span>, <span class="fl">2</span>, <span class="va">ColToHex</span><span class="op">)</span>
<span class="va">spec3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">spec3</span>, <span class="fl">2</span>, <span class="va">ColToHex</span><span class="op">)</span></code></pre></div>
<p>Now we can plot this information on a ternary diagram.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">'Ternary'</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/layout.html">layout</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>, widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>
  <span class="va">oldPar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/TernaryPlot.html">TernaryPlot</a></span><span class="op">(</span>alab <span class="op">=</span> <span class="st">'Information absent in degraded tree'</span>, 
              blab <span class="op">=</span> <span class="st">'\n\nCorrect information in degraded tree'</span>, 
              clab <span class="op">=</span> <span class="st">'Misinformation in degraded tree'</span>,
              point <span class="op">=</span> <span class="st">'left'</span>, clockwise <span class="op">=</span> <span class="cn">FALSE</span>, grid.minor.lines <span class="op">=</span> <span class="fl">0</span>,
              axis.labels <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">10</span> <span class="op">/</span> <span class="fl">10</span><span class="op">)</span>
  
  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/TernaryPlot.html">HorizontalGrid</a></span><span class="op">(</span><span class="op">)</span>
  
  <span class="va">coords1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">unresolved1</span>, <span class="va">correct1</span>, <span class="va">incorrect1</span><span class="op">)</span>
  <span class="va">coords3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">unresolved3</span>, <span class="va">correct3</span>, <span class="va">incorrect3</span><span class="op">)</span>
  
  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/ColourTernary.html">ColourTernary</a></span><span class="op">(</span><span class="fu"><a href="https://ms609.github.io/Ternary/reference/TernaryPointValues.html">TernaryDensity</a></span><span class="op">(</span><span class="va">coords1</span>, resolution <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>, spectrum <span class="op">=</span> <span class="va">spec1</span><span class="op">)</span>
  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/ColourTernary.html">ColourTernary</a></span><span class="op">(</span><span class="fu"><a href="https://ms609.github.io/Ternary/reference/TernaryPointValues.html">TernaryDensity</a></span><span class="op">(</span><span class="va">coords3</span>, resolution <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>, spectrum <span class="op">=</span> <span class="va">spec3</span><span class="op">)</span>
  
  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/TernaryDensityContour.html">TernaryDensityContour</a></span><span class="op">(</span><span class="va">coords3</span>, col <span class="op">=</span> <span class="va">col3</span>, nlevels <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>
  <span class="fu"><a href="https://ms609.github.io/Ternary/reference/TernaryDensityContour.html">TernaryDensityContour</a></span><span class="op">(</span><span class="va">coords1</span>, col <span class="op">=</span> <span class="va">col1</span>, nlevels <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>
  
  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">'kdensity'</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/JonasMoss/kdensity">'kdensity'</a></span><span class="op">)</span>
    <span class="va">HorizontalKDE</span> <span class="op">&lt;-</span> <span class="kw">function</span> <span class="op">(</span><span class="va">dat</span>, <span class="va">col</span>, <span class="va">add</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span>
      <span class="va">lty</span> <span class="op">&lt;-</span> <span class="fl">1</span>
      <span class="va">lwd</span> <span class="op">&lt;-</span> <span class="fl">2</span>
      <span class="va">kde</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/kdensity/man/kdensity.html">kdensity</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span>
      <span class="va">kdeRange</span> <span class="op">&lt;-</span> <span class="fu">kdensity</span><span class="fu">:::</span><span class="fu"><a href="https://rdrr.io/pkg/kdensity/man/get_range.html">get_range</a></span><span class="op">(</span><span class="va">kde</span><span class="op">)</span>
      <span class="kw">if</span> <span class="op">(</span><span class="va">add</span><span class="op">)</span> <span class="op">{</span>
        <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fu">kde</span><span class="op">(</span><span class="va">kdeRange</span><span class="op">)</span>, <span class="va">kdeRange</span>, col <span class="op">=</span> <span class="va">col</span>, lty <span class="op">=</span> <span class="va">lty</span>, lwd <span class="op">=</span> <span class="va">lwd</span><span class="op">)</span>
      <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span>
        <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu">kde</span><span class="op">(</span><span class="va">kdeRange</span><span class="op">)</span>, <span class="va">kdeRange</span>, col <span class="op">=</span> <span class="va">col</span>, lty <span class="op">=</span> <span class="va">lty</span>, lwd <span class="op">=</span> <span class="va">lwd</span>, 
             ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">''</span>, axes <span class="op">=</span> <span class="cn">FALSE</span>, type <span class="op">=</span> <span class="st">'l'</span><span class="op">)</span>
      <span class="op">}</span>
      <span class="co"># abline(h = 0:10 / 10) # Useful for confirming alignment</span>
    <span class="op">}</span>
  
    <span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1.8</span>, <span class="fl">0</span>, <span class="fl">1.8</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="co"># align plot limits with ternary plot</span>
    <span class="fu">HorizontalKDE</span><span class="op">(</span><span class="va">correct1</span> <span class="op">/</span> <span class="va">infoInTree1</span>, <span class="va">col1</span>, add <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
    <span class="fu">HorizontalKDE</span><span class="op">(</span><span class="va">correct3</span> <span class="op">/</span> <span class="va">infoInTree3</span>, <span class="va">col3</span>, add <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/graphics/mtext.html">mtext</a></span><span class="op">(</span><span class="st">'\u2192 Normalized tree quality \u2192'</span>, <span class="fl">2</span><span class="op">)</span>
  <span class="op">}</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span><span class="va">oldPar</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="using-distances_files/figure-html/unnamed-chunk-22-1.png" width="70%" style="display: block; margin: auto;"></p>
<p>In the ternary plot, the vertical direction corresponds to the normalized tree quality, as depicted in the accompanying histogram.</p>
</div>
</div>
<div id="what-next" class="section level2">
<h2 class="hasAnchor">
<a href="#what-next" class="anchor"></a>What next?</h2>
<p>You may wish to:</p>
<ul>
<li><p>Explore the <a href="https://ms609.github.io/Ternary/">Ternary package</a></p></li>
<li><p><a href="https://ms609.github.io/TreeDistData/articles/09-expected-similarity.html">Interpret tree distance metrics</a></p></li>
<li><p>Review <a href="https://ms609.github.io/TreeDist/index.html">available distance measures</a> and the corresponding <a href="https://ms609.github.io/TreeDist/reference/index.html#section-tree-distance-measures">TreeDist functions</a></p></li>
<li><p>Construct <a href="treespace.html">tree spaces</a> to visualize landscapes of phylogenetic trees</p></li>
</ul>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-Nye2006">
<p>Nye, T. M. W., Liò, P., &amp; Gilks, W. R. (2006). A novel algorithm and web-based tool for comparing two alternative phylogenetic trees. <em>Bioinformatics</em>, <em>22</em>(1), 117–119. doi:<a href="https://doi.org/10.1093/bioinformatics/bti720">10.1093/bioinformatics/bti720</a></p>
</div>
<div id="ref-Smith2019">
<p>Smith, M. R. (2019). Bayesian and parsimony approaches reconstruct informative trees from simulated morphological datasets. <em>Biology Letters</em>, <em>15</em>(2), 20180632. doi:<a href="https://doi.org/10.1098/rsbl.2018.0632">10.1098/rsbl.2018.0632</a></p>
</div>
<div id="ref-SmithDist">
<p>Smith, M. R. (2020). Information theoretic Generalized Robinson-Foulds metrics for comparing phylogenetic trees. <em>Bioinformatics</em>, <em>36</em>(20), 5007–5013. doi:<a href="https://doi.org/10.1093/bioinformatics/btaa614">10.1093/bioinformatics/btaa614</a></p>
</div>
<div id="ref-Vinh2010">
<p>Vinh, N. X., Epps, J., &amp; Bailey, J. (2010). Information theoretic measures for clusterings comparison: variants, properties, normalization and correction for chance. <em>Journal of Machine Learning Research</em>, <em>11</em>, 2837–2854. doi:<a href="https://doi.org/10.1145/1553374.1553511">10.1145/1553374.1553511</a></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://community.dur.ac.uk/martin.smith/">Martin R. Smith</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '8bf44d312479af73720c5ca5fcc12eba',
    indexName: 'treedist',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
