<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Information-based generalized Robinson–Foulds distances — TreeDistance • TreeDist</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Information-based generalized Robinson–Foulds distances — TreeDistance"><meta name="description" content="Calculate tree similarity and distance measures based on the amount of
phylogenetic or clustering information that two trees hold in common, as
proposed in Smith (2020)."><meta property="og:description" content="Calculate tree similarity and distance measures based on the amount of
phylogenetic or clustering information that two trees hold in common, as
proposed in Smith (2020)."><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-BTKEG66D3F"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-BTKEG66D3F');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TreeDist</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">2.10.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home fa-lg"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/Using-TreeDist.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Function reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tree distance analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/Using-TreeDist.html">Calculate tree similarity with 'TreeDist'</a></li>
    <li><a class="dropdown-item" href="../articles/using-distances.html">Contextualizing tree distances</a></li>
    <li><a class="dropdown-item" href="../articles/different-leaves.html">Trees with different leaves</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tree space analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/treespace.html">Tree space analysis</a></li>
    <li><a class="dropdown-item" href="../articles/landscapes.html">Analysing landscapes of phylogenetic trees</a></li>
    <li><a class="dropdown-item" href="../articles/compare-treesets.html">Comparing sets of trees from different analyses</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tree distance introductions</h6></li>
    <li><a class="dropdown-item" href="../articles/information.html">Comparing splits using information theory</a></li>
    <li><a class="dropdown-item" href="../articles/Robinson-Foulds.html">Extending the Robinson-Foulds metric</a></li>
    <li><a class="dropdown-item" href="../articles/Generalized-RF.html">Generalized Robinson-Foulds distances</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ms609/TreeDist"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Information-based generalized Robinson–Foulds distances</h1>
      <small class="dont-index">Source: <a href="https://github.com/ms609/TreeDist/blob/main/R/tree_distance_info.R" class="external-link"><code>R/tree_distance_info.R</code></a>, <a href="https://github.com/ms609/TreeDist/blob/main/R/tree_distance_msi.R" class="external-link"><code>R/tree_distance_msi.R</code></a></small>
      <div class="d-none name"><code>TreeDistance.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Calculate tree similarity and distance measures based on the amount of
phylogenetic or clustering information that two trees hold in common, as
proposed in Smith (2020).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">TreeDistance</span><span class="op">(</span><span class="va">tree1</span>, tree2 <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">SharedPhylogeneticInfo</span><span class="op">(</span></span>
<span>  <span class="va">tree1</span>,</span>
<span>  tree2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normalize <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reportMatching <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  diag <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">DifferentPhylogeneticInfo</span><span class="op">(</span></span>
<span>  <span class="va">tree1</span>,</span>
<span>  tree2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normalize <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reportMatching <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">PhylogeneticInfoDistance</span><span class="op">(</span></span>
<span>  <span class="va">tree1</span>,</span>
<span>  tree2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normalize <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reportMatching <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ClusteringInfoDistance</span><span class="op">(</span></span>
<span>  <span class="va">tree1</span>,</span>
<span>  tree2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normalize <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reportMatching <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ExpectedVariation</span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>, samples <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">MutualClusteringInfo</span><span class="op">(</span></span>
<span>  <span class="va">tree1</span>,</span>
<span>  tree2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normalize <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reportMatching <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  diag <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">SharedPhylogeneticInfoSplits</span><span class="op">(</span></span>
<span>  <span class="va">splits1</span>,</span>
<span>  <span class="va">splits2</span>,</span>
<span>  nTip <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">splits1</span>, <span class="st">"nTip"</span><span class="op">)</span>,</span>
<span>  reportMatching <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">MutualClusteringInfoSplits</span><span class="op">(</span></span>
<span>  <span class="va">splits1</span>,</span>
<span>  <span class="va">splits2</span>,</span>
<span>  nTip <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">splits1</span>, <span class="st">"nTip"</span><span class="op">)</span>,</span>
<span>  reportMatching <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">MatchingSplitInfo</span><span class="op">(</span></span>
<span>  <span class="va">tree1</span>,</span>
<span>  tree2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normalize <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reportMatching <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  diag <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">MatchingSplitInfoDistance</span><span class="op">(</span></span>
<span>  <span class="va">tree1</span>,</span>
<span>  tree2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normalize <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reportMatching <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">MatchingSplitInfoSplits</span><span class="op">(</span></span>
<span>  <span class="va">splits1</span>,</span>
<span>  <span class="va">splits2</span>,</span>
<span>  nTip <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">splits1</span>, <span class="st">"nTip"</span><span class="op">)</span>,</span>
<span>  reportMatching <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-tree-tree-">tree1, tree2<a class="anchor" aria-label="anchor" href="#arg-tree-tree-"></a></dt>
<dd><p>Trees of class <code>phylo</code>, with leaves labelled identically,
or lists of such trees to undergo pairwise comparison.  Where implemented,
<code>tree2 = NULL</code> will compute distances between each pair of trees in the list
<code>tree1</code> using a fast algorithm based on
Day (1985)
.</p></dd>


<dt id="arg-normalize">normalize<a class="anchor" aria-label="anchor" href="#arg-normalize"></a></dt>
<dd><p>If a numeric value is provided, this will be used as a
maximum value against which to rescale results.
If <code>TRUE</code>, results will be rescaled against a maximum value calculated from
the specified tree sizes and topology, as specified in the "Normalization"
section below.
If <code>FALSE</code>, results will not be rescaled.</p></dd>


<dt id="arg-reportmatching">reportMatching<a class="anchor" aria-label="anchor" href="#arg-reportmatching"></a></dt>
<dd><p>Logical specifying whether to return the clade
matchings as an attribute of the score.</p></dd>


<dt id="arg-diag">diag<a class="anchor" aria-label="anchor" href="#arg-diag"></a></dt>
<dd><p>Logical specifying whether to return similarities along the
diagonal, i.e. of each tree with itself.  Applies only if <code>tree2</code> is
a list identical to <code>tree1</code>, or <code>NULL</code>.</p></dd>


<dt id="arg-samples">samples<a class="anchor" aria-label="anchor" href="#arg-samples"></a></dt>
<dd><p>Integer specifying how many samplings to obtain;
accuracy of estimate increases with <code>sqrt(samples)</code>.</p></dd>


<dt id="arg-splits-splits-">splits1, splits2<a class="anchor" aria-label="anchor" href="#arg-splits-splits-"></a></dt>
<dd><p>Logical matrices where each row corresponds to a leaf,
either listed in the same order or bearing identical names (in any sequence),
and each column corresponds to a split, such that each leaf is identified as
a member of the ingroup (<code>TRUE</code>) or outgroup (<code>FALSE</code>) of the respective
split.</p></dd>


<dt id="arg-ntip">nTip<a class="anchor" aria-label="anchor" href="#arg-ntip"></a></dt>
<dd><p>(Optional) Integer specifying the number of leaves in each split.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>If <code>reportMatching = FALSE</code>, the functions return a numeric
vector specifying the requested similarities or differences.</p>
<p>If <code>reportMatching = TRUE</code>, the functions additionally return an integer
vector listing the index of the split in <code>tree2</code> that is matched with
each split in <code>tree1</code> in the optimal matching.
Unmatched splits are denoted <code>NA</code>.
Use <code><a href="VisualizeMatching.html">VisualizeMatching()</a></code> to plot the optimal matching.</p>
<p><code>TreeDistance()</code> simply returns the clustering information distance (it is
an alias of <code>ClusteringInfoDistance()</code>).</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><a href="https://ms609.github.io/TreeDist/articles/Robinson-Foulds.html#generalized-robinson-foulds-distances">Generalized Robinson–Foulds distances</a>
calculate tree similarity by finding an
optimal matching that the similarity between a split on one tree
and its pair on a second, considering all possible ways to pair splits
between trees (including leaving a split unpaired).</p>
<p>The methods implemented here use the concepts of
<a href="https://ms609.github.io/TreeDist/articles/information.html">entropy and information</a>
(MacKay 2003)
 to assign a similarity score between each
pair of splits.</p>
<p>The returned tree similarity measures state the amount of information,
in bits, that the splits in two trees hold in common
when they are optimally matched, following
Smith (2020)
.
The complementary tree distance measures state how much information is
different in the splits of two trees, under an optimal matching.
Where trees contain different tips, tips present in one tree but not the
other are removed before each comparison (as by definition, the trees neither
hold information in common nor differ regarding these tips).</p>
    </div>
    <div class="section level2">
    <h2 id="concepts-of-information">Concepts of information<a class="anchor" aria-label="anchor" href="#concepts-of-information"></a></h2>
    <p>The phylogenetic (Shannon) information content and entropy of a split are
defined in
<a href="https://ms609.github.io/TreeDist/articles/information.html">a separate vignette</a>.</p>
<p>Using the mutual (clustering) information
(Meila 2007; Vinh et al. 2010)
 of two splits to quantify their
similarity gives rise to the Mutual Clustering Information measure
(<code>MutualClusteringInfo()</code>, <code>MutualClusteringInfoSplits()</code>);
the entropy distance gives the Clustering Information Distance
(<code>ClusteringInfoDistance()</code>).
This approach is optimal in many regards, and is implemented with
normalization in the convenience function <code>TreeDistance()</code>.</p>
<p>Using the amount of phylogenetic information common to two splits to measure
their similarity gives rise to the Shared Phylogenetic Information similarity
measure (<code>SharedPhylogeneticInfo()</code>, <code>SharedPhylogeneticInfoSplits()</code>).
The amount of information distinct to
each of a pair of splits provides the complementary Different Phylogenetic
Information distance metric (<code>DifferentPhylogeneticInfo()</code>).</p>
<p>The Matching Split Information measure (<code>MatchingSplitInfo()</code>,
<code>MatchingSplitInfoSplits()</code>) defines the similarity between a pair of
splits as the phylogenetic information content of the most informative
split that is consistent with both input splits; <code>MatchingSplitInfoDistance()</code>
is the corresponding measure of tree difference.
(<a href="https://ms609.github.io/TreeDist/articles/Generalized-RF.html">More information here</a>.)</p>
    </div>
    <div class="section level2">
    <h2 id="conversion-to-distances">Conversion to distances<a class="anchor" aria-label="anchor" href="#conversion-to-distances"></a></h2>
    <p>To convert similarity measures to distances, it is necessary to
subtract the similarity score from a maximum value.  In order to generate
distance <em>metrics</em>, these functions subtract the similarity twice from the
total information content (SPI, MSI) or entropy (MCI) of all the splits in
both trees (Smith 2020)
.</p>
    </div>
    <div class="section level2">
    <h2 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a></h2>
    <p>If <code>normalize = TRUE</code>, then results will be rescaled such that distance
ranges from zero to (in principle) one.
The maximum <strong>distance</strong> is the sum of the information content or entropy of
each split in each tree; the maximum <strong>similarity</strong> is half this value.
(See Vinh <em>et al.</em> (2010, table 3) and
Smith (2020)
 for
alternative normalization possibilities.)</p>
<p>Note that a distance value of one (= similarity of zero) will seldom be
achieved, as even the most different trees exhibit some similarity.
It may thus be helpful to rescale the normalized value such that the
<em>expected</em> distance between a random pair of trees equals one.  This can
be calculated with <code>ExpectedVariation()</code>; or see package
'<a href="https://ms609.github.io/TreeDistData/reference/randomTreeDistances.html">TreeDistData</a>'
for a compilation of expected values under different metrics for trees with
up to 200 leaves.</p>
<p>Alternatively, use <code>normalize = </code><code><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmax</a></code> or <code><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmin</a></code> to scale against the
information content or entropy of all splits in the most (<code>pmax</code>) or
least (<code>pmin</code>) informative tree in each pair.
To calculate the relative similarity against a reference tree that is known
to be "correct", use <code>normalize = SplitwiseInfo(trueTree)</code> (SPI, MSI) or
<code>ClusteringEntropy(trueTree)</code> (MCI).
For worked examples, see the internal function <code><a href="NormalizeInfo.html">NormalizeInfo()</a></code>, which is
called from distance functions with the parameter <code>how = normalize</code>.
.</p>
    </div>
    <div class="section level2">
    <h2 id="distances-between-large-trees">Distances between large trees<a class="anchor" aria-label="anchor" href="#distances-between-large-trees"></a></h2>
    <p>To balance memory demands and runtime with flexibility, these functions are
implemented for trees with up to 2048 leaves.
To analyse trees with up to 8192 leaves, you will need to a modified version
of the package:
<code>install.packages("BigTreeDist", repos = "https://ms609.github.io/packages/")</code>
Use <code><a href="https://rdrr.io/r/base/library.html" class="external-link">library("BigTreeDist")</a></code> <em>instead</em> of <code><a href="https://ms609.github.io/TreeDist/">library("TreeDist")</a></code> to load
the modified package – or prefix functions with the package name, e.g.
<code>BigTreeDist::TreeDistance()</code>.</p>
<p>As an alternative download method,
uninstall <span class="pkg">TreeDist</span> and <span class="pkg">TreeTools</span> using
<code><a href="https://rdrr.io/r/utils/remove.packages.html" class="external-link">remove.packages()</a></code>, then use
<code>devtools::install_github("ms609/TreeTools", ref = "more-leaves")</code>
to install the modified <span class="pkg">TreeTools</span> package; then,
install <span class="pkg">TreeDist</span> using
<code>devtools::install_github("ms609/TreeDist", ref = "more-leaves")</code>.
(<span class="pkg">TreeDist</span> will need building from source <em>after</em> the modified
<span class="pkg">TreeTools</span> package has been installed, as its code links to values
set in the TreeTools source code.)</p>
<p>Trees with over 8192 leaves require further modification of the source code,
which the maintainer plans to attempt in the future; please <a href="https://github.com/ms609/TreeTools/issues/141" class="external-link">comment on GitHub</a> if you would find this useful.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Day WHE (1985).
“Optimal algorithms for comparing trees with labeled leaves.”
<em>Journal of Classification</em>, <b>2</b>(1), 7–28.
<a href="https://doi.org/10.1007/BF01908061" class="external-link">doi:10.1007/BF01908061</a>
.<br><br> MacKay DJC (2003).
<em>Information Theory, Inference, and Learning Algorithms</em>.
Cambridge University Press, Cambridge.
<a href="https://www.inference.org.uk/itprnn/book.pdf" class="external-link">https://www.inference.org.uk/itprnn/book.pdf</a>.<br><br> Meila M (2007).
“Comparing clusterings—an information based distance.”
<em>Journal of Multivariate Analysis</em>, <b>98</b>(5), 873–895.
<a href="https://doi.org/10.1016/j.jmva.2006.11.013" class="external-link">doi:10.1016/j.jmva.2006.11.013</a>
.<br><br> Smith MR (2020).
“Information theoretic Generalized Robinson-Foulds metrics for comparing phylogenetic trees.”
<em>Bioinformatics</em>, <b>36</b>(20), 5007–5013.
<a href="https://doi.org/10.1093/bioinformatics/btaa614" class="external-link">doi:10.1093/bioinformatics/btaa614</a>
.<br><br> Vinh NX, Epps J, Bailey J (2010).
“Information theoretic measures for clusterings comparison: variants, properties, normalization and correction for chance.”
<em>Journal of Machine Learning Research</em>, <b>11</b>, 2837–2854.
<a href="https://doi.org/10.1145/1553374.1553511" class="external-link">doi:10.1145/1553374.1553511</a>
.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other tree distances:
<code><a href="JaccardRobinsonFoulds.html">JaccardRobinsonFoulds</a>()</code>,
<code><a href="KendallColijn.html">KendallColijn</a>()</code>,
<code><a href="MASTSize.html">MASTSize</a>()</code>,
<code><a href="MatchingSplitDistance.html">MatchingSplitDistance</a>()</code>,
<code><a href="NNIDist.html">NNIDist</a>()</code>,
<code><a href="NyeSimilarity.html">NyeSimilarity</a>()</code>,
<code><a href="PathDist.html">PathDist</a>()</code>,
<code><a href="Robinson-Foulds.html">Robinson-Foulds</a></code>,
<code><a href="SPRDist.html">SPRDist</a>()</code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p><a href="https://orcid.org/0000-0001-5660-1727" class="external-link">Martin R. Smith</a>
(<a href="mailto:martin.smith@durham.ac.uk">martin.smith@durham.ac.uk</a>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">tree1</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span>text<span class="op">=</span><span class="st">"((((a, b), c), d), (e, (f, (g, h))));"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tree2</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span>text<span class="op">=</span><span class="st">"(((a, b), (c, d)), ((e, f), (g, h)));"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tree3</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span>text<span class="op">=</span><span class="st">"((((h, b), c), d), (e, (f, (g, a))));"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Best possible score is obtained by matching a tree with itself</span></span></span>
<span class="r-in"><span><span class="fu">DifferentPhylogeneticInfo</span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree1</span><span class="op">)</span> <span class="co"># 0, by definition</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0</span>
<span class="r-in"><span><span class="fu">SharedPhylogeneticInfo</span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 22.53747</span>
<span class="r-in"><span><span class="fu"><a href="TreeInfo.html">SplitwiseInfo</a></span><span class="op">(</span><span class="va">tree1</span><span class="op">)</span> <span class="co"># Maximum shared phylogenetic information</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 22.53747</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Best possible score is a function of tree shape; the splits within</span></span></span>
<span class="r-in"><span><span class="co"># balanced trees are more independent and thus contain less information</span></span></span>
<span class="r-in"><span><span class="fu"><a href="TreeInfo.html">SplitwiseInfo</a></span><span class="op">(</span><span class="va">tree2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 19.36755</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># How similar are two trees?</span></span></span>
<span class="r-in"><span><span class="fu">SharedPhylogeneticInfo</span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span><span class="op">)</span> <span class="co"># Amount of phylogenetic information in common</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 13.75284</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="fu">SharedPhylogeneticInfo</span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>, reportMatching <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="st">"matching"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1 4 2 3 5</span>
<span class="r-in"><span><span class="fu"><a href="VisualizeMatching.html">VisualizeMatching</a></span><span class="op">(</span><span class="va">SharedPhylogeneticInfo</span>, <span class="va">tree1</span>, <span class="va">tree2</span><span class="op">)</span> <span class="co"># Which clades are matched?</span></span></span>
<span class="r-plt img"><img src="TreeDistance-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">DifferentPhylogeneticInfo</span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span><span class="op">)</span> <span class="co"># Distance measure</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 14.39934</span>
<span class="r-in"><span><span class="fu">DifferentPhylogeneticInfo</span><span class="op">(</span><span class="va">tree2</span>, <span class="va">tree1</span><span class="op">)</span> <span class="co"># The metric is symmetric</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 14.39934</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Are they more similar than two trees of this shape would be by chance?</span></span></span>
<span class="r-in"><span><span class="fu">ExpectedVariation</span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree2</span>, sample<span class="op">=</span><span class="fl">12</span><span class="op">)</span><span class="op">[</span><span class="st">"DifferentPhylogeneticInfo"</span>, <span class="st">"Estimate"</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 34.5706</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Every split in tree1 conflicts with every split in tree3</span></span></span>
<span class="r-in"><span><span class="co"># Pairs of conflicting splits contain clustering, but not phylogenetic, </span></span></span>
<span class="r-in"><span><span class="co"># information</span></span></span>
<span class="r-in"><span><span class="fu">SharedPhylogeneticInfo</span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree3</span><span class="op">)</span> <span class="co"># = 0</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0</span>
<span class="r-in"><span><span class="fu">MutualClusteringInfo</span><span class="op">(</span><span class="va">tree1</span>, <span class="va">tree3</span><span class="op">)</span> <span class="co"># &gt; 0</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.6539805</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Distance functions internally convert trees to Splits objects.</span></span></span>
<span class="r-in"><span><span class="co"># Pre-conversion can reduce run time if the same trees will feature in</span></span></span>
<span class="r-in"><span><span class="co"># multiple comparisons</span></span></span>
<span class="r-in"><span><span class="va">splits1</span> <span class="op">&lt;-</span> <span class="fu">TreeTools</span><span class="fu">::</span><span class="fu"><a href="https://ms609.github.io/TreeTools/reference/Splits.html" class="external-link">as.Splits</a></span><span class="op">(</span><span class="va">tree1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">splits2</span> <span class="op">&lt;-</span> <span class="fu">TreeTools</span><span class="fu">::</span><span class="fu"><a href="https://ms609.github.io/TreeTools/reference/Splits.html" class="external-link">as.Splits</a></span><span class="op">(</span><span class="va">tree2</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">SharedPhylogeneticInfoSplits</span><span class="op">(</span><span class="va">splits1</span>, <span class="va">splits2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 13.75284</span>
<span class="r-in"><span><span class="fu">MatchingSplitInfoSplits</span><span class="op">(</span><span class="va">splits1</span>, <span class="va">splits2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 17.09254</span>
<span class="r-in"><span><span class="fu">MutualClusteringInfoSplits</span><span class="op">(</span><span class="va">splits1</span>, <span class="va">splits2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 3.031424</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://smithlabdurham.github.io/" class="external-link">Martin R. Smith</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.9000.</p>
</div>

    </footer></div>





  </body></html>

