<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Generalized Robinson-Foulds distances â€¢ TreeDist</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Generalized Robinson-Foulds distances">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-156666971-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156666971-1');
</script>
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TreeDist</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/Using-TreeDist.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Using-TreeDist.html">Using TreeDist to calculate tree similarity</a>
    </li>
    <li>
      <a href="../articles/using-distances.html">Contextualizing tree distances</a>
    </li>
    <li>
      <a href="../articles/Robinson-Foulds.html">Robinson-Foulds metric</a>
    </li>
    <li>
      <a href="../articles/Generalized-RF.html">Generalized Robinson-Foulds distances</a>
    </li>
    <li>
      <a href="../articles/information.html">Concepts of information</a>
    </li>
    <li>
      <a href="../articles/Tree-distance-metric-evaluation.html">Evaluating tree distance metrics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ms609/TreeDist">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Generalized Robinson-Foulds distances</h1>
                        <h4 class="author">Martin R. Smith</h4>
            
            <h4 class="date">2020-01-28</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ms609/TreeDist/blob/master/vignettes/Generalized-RF.Rmd"><code>vignettes/Generalized-RF.Rmd</code></a></small>
      <div class="hidden name"><code>Generalized-RF.Rmd</code></div>

    </div>

    
    
<p>This document outlines the similarity measures employed by the generalized Robinson-Foulds distances implemented in this package.</p>
<p>Generalized RF distances are introduced <a href="Robinson-Foulds.html">elsewhere</a>; before you read further, you may also wish to revisit how to <a href="Using-TreeDist.html">use the TreeDist package</a>, and relevant <a href="information.html">principles of information theory</a>.</p>
<div id="shared-phylogenetic-information" class="section level2">
<h2 class="hasAnchor">
<a href="#shared-phylogenetic-information" class="anchor"></a>Shared phylogenetic information</h2>
<p>Under the shared phylogenetic information tree distance measure <span class="citation">(Smith, 2020)</span>, pairs of splits are assigned a similarity score that corresponds to the amount of phylogenetic information <span class="citation">(<em>sensu</em> Steel &amp; Penny, 2006)</span> that they share in common (see <a href="information.html">separate vignette</a>).</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(SharedPhylogeneticInfo, tree1, tree2, </a>
<a class="sourceLine" id="cb1-2" data-line-number="2">                  <span class="dt">Plot=</span>TreeDistPlot, <span class="dt">matchZeros =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-1-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="../reference/TreeDistance.html">SharedPhylogeneticInfo</a></span>(tree1, tree2)</a></code></pre></div>
<pre><code>## [1] 12.32188</code></pre>
<div id="the-problem-with-arboreal-matchings" class="section level3">
<h3 class="hasAnchor">
<a href="#the-problem-with-arboreal-matchings" class="anchor"></a>The problem with arboreal matchings</h3>
<p>Shared phylogenetic information is an arboreal similarity score: that is to say, it assigns zero similarity to splits that cannot both occur on a single tree.<br>
This leads to problematic behaviour in certain cases: for example, when the position of two distant taxa is swapped, similarities within a tree can be obscured.<br>
The two trees below are identical but for the position of <em>A</em> and <em>J</em>, yet receive zero similarity under arboreal metrics.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(SharedPhylogeneticInfo, <span class="dt">matchZeros =</span> <span class="ot">FALSE</span>, AtoJ, AotJ,</a>
<a class="sourceLine" id="cb4-2" data-line-number="2">                  <span class="dt">Plot=</span>TreeDistPlot, <span class="dt">leaveRoom =</span> <span class="ot">FALSE</span>, <span class="dt">prune=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">5</span>, <span class="dv">18</span>))</a></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-2-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="mutual-clustering-information" class="section level2">
<h2 class="hasAnchor">
<a href="#mutual-clustering-information" class="anchor"></a>Mutual clustering information</h2>
<p>Scoring each pair of splits according to their mutual clustering information <span class="citation">(Smith, 2020)</span> (see <a href="information.html">separate vignette</a>) results in a non-arboreal information-based tree distance metric that, as desired, assigns a non-zero similarity to the case that received zero similarity under an arboreal matching:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(MutualClusteringInfo, AtoJ, AotJ, <span class="dt">matchZeros =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb5-2" data-line-number="2">                  <span class="dt">Plot=</span>TreeDistPlot, <span class="dt">leaveRoom =</span> <span class="ot">FALSE</span>, <span class="dt">prune=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">5</span>, <span class="dv">18</span>))</a></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-3-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw"><a href="../reference/TreeDistance.html">MutualClusteringInfo</a></span>(AtoJ, AotJ)</a></code></pre></div>
<pre><code>## [1] 1.348339</code></pre>
<p>Because no pair of non-trivial splits has zero mutual clustering information, even a dissimilar matching (such as <code>HI|ABCDEFG</code> =&gt; <code>EI|ABCDFGH</code> below) is preferred to leaving a split unpaired.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(MutualClusteringInfo, tree1, tree2, </a>
<a class="sourceLine" id="cb8-2" data-line-number="2">                  <span class="dt">Plot=</span>TreeDistPlot, <span class="dt">leaveRoom=</span><span class="ot">FALSE</span>, <span class="dt">matchZeros =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="nye-et-al--tree-similarity-metric" class="section level2">
<h2 class="hasAnchor">
<a href="#nye-et-al--tree-similarity-metric" class="anchor"></a>Nye <em>et al.</em> tree similarity metric</h2>
<p>The Nye <em>et al</em>. <span class="citation">(2006)</span> tree similarity metric scores pairs by considering the elements held in common between subsets of each split.</p>
<p>Consider a pair of splits <code>ABCDEF|GHIJ</code> and <code>ABCDEIJ|FGH</code>. These can be aligned thus:</p>
<pre><code>ABCDEF  | GHIJ
ABCDE IJ|FGH</code></pre>
<p>The first pair of subsets, <code>ABCDEF</code> and <code>ABCDEIJ</code>, have five elements in common (<code>ABCDE</code>), and together encompass eight elements (<code>ABCDEFIJ</code>). Their <em>subset score</em> is thus 5/8.</p>
<p>The second pair of subsets, <code>GHIJ</code> and <code>FGH</code>, have two elements (<code>GH</code>) in common, of the five total (<code>FGHIJ</code>), and hence receive a subset score of 2/5.</p>
<p>This split alignment then receives an <em>alignment score</em> corresponding to the lower of the two subset scores, 2/5.</p>
<p>We must now consider the other alignment of this pair of splits,</p>
<pre><code>ABCDEF  |     GHIJ
     FGH|ABCDE  IJ</code></pre>
<p>This yields subset scores of 1/8 and 2/9, and thus has an alignment score of 1/8. This alignment gives a lower score than the other, so is disregarded. The pair of splits is allocated a similarity score corresponding to the better alignment: 2/5.</p>
<p>As such, splits that match exactly will receive a similarity score of 1, in a manner analogous to the Robinson-Foulds distance. (This is despite the fact that some splits are more likely to match than others.) It is not possible for a pair of splits to receive a zero similarity score.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(NyeTreeSimilarity, tree1, tree2, </a>
<a class="sourceLine" id="cb11-2" data-line-number="2">                  <span class="dt">Plot=</span>TreeDistPlot, <span class="dt">leaveRoom=</span><span class="ot">FALSE</span>, <span class="dt">matchZeros =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-5-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw"><a href="../reference/NyeTreeSimilarity.html">NyeTreeSimilarity</a></span>(tree1, tree2, <span class="dt">normalize=</span><span class="ot">FALSE</span>)</a></code></pre></div>
<pre><code>## [1] 3.5</code></pre>
<div id="jaccard-robinson-foulds-metric" class="section level3">
<h3 class="hasAnchor">
<a href="#jaccard-robinson-foulds-metric" class="anchor"></a>Jaccard-Robinson-Foulds metric</h3>
<p>BÃ¶cker <em>et al</em>. <span class="citation">(2013)</span> propose expressly prohibiting contradictory pairings, i.e.Â pairs of splits that could never occur in the same tree. Such an â€˜arborealâ€™ matching would preclude the pairing of <code>ABCFGHI|DE</code> with <code>ABCDFGH|EI</code>:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw"><a href="../reference/JaccardRobinsonFoulds.html">JaccardRobinsonFoulds</a></span>(tree1, tree2, <span class="dt">similarity =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] 6.333333</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="kw"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(JaccardRobinsonFoulds, tree1, tree2,</a>
<a class="sourceLine" id="cb16-2" data-line-number="2">                  <span class="dt">Plot=</span>TreeDistPlot, <span class="dt">leaveRoom=</span><span class="ot">FALSE</span>, <span class="dt">matchZeros =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;"></p>
<p>BÃ¶cker <em>et al</em>. <span class="citation">(2013)</span> also suggest raising the split similarity score defined above to an arbitrary exponent. The similarity score is also doubled. By this means, the metric can be made to converge to the Robinson-Foulds metric as the exponent grows towards infinity.</p>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-7-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="matching-split-distance" class="section level2">
<h2 class="hasAnchor">
<a href="#matching-split-distance" class="anchor"></a>Matching Split Distance</h2>
<p>Bogdanowicz &amp; Giaro <span class="citation">(2012)</span> propose an alternative distance, which they term the Matching Split Distance.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw"><a href="../reference/MatchingSplitDistance.html">MatchingSplitDistance</a></span>(tree1, tree2)</a></code></pre></div>
<pre><code>## [1] 5</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="kw"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(MatchingSplitDistance, tree1, tree2,</a>
<a class="sourceLine" id="cb19-2" data-line-number="2">                  <span class="dt">Plot=</span>TreeDistPlot, <span class="dt">leaveRoom=</span><span class="ot">FALSE</span>, <span class="dt">matchZeros =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Note that the visualization shows the difference, rather than the similarity, between splits.</p>
<p>Similar to the Nye <em>et al</em>. similarity metric, this method compares the subsets implied by a pair of splits. Here, the relevant quantity is the number of elements that must be moved from one subset to another in order to make the two splits identical. With the pair of splits</p>
<pre><code>ABCDEF  | GHIJ
ABCDE IJ|FGH</code></pre>
<p>three leaves (<em>F</em>, <em>I</em> and <em>J</em>) must be moved before the splits are identical; as such, the pair of splits are assigned a difference score of three.<br>
Formally, where <span class="math inline">\(S_i\)</span> splits <span class="math inline">\(n\)</span> leaves into bipartitions <span class="math inline">\(A_i\)</span> and <span class="math inline">\(B_i\)</span>, the difference score is calculated by</p>
<p><span class="math inline">\(n - m\)</span></p>
<p>where <span class="math inline">\(m\)</span> counts the number of leaves that already match, and is defined as</p>
<p><span class="math inline">\(m = \max\{|A_1 \cap A_2| + |B_1 \cap B_2|, |A_1 \cap B_2| + |B_1 \cap A_2|\}\)</span></p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="kw"><a href="../reference/MatchingSplitDistance.html">MatchingSplitDistance</a></span>(<span class="kw">read.tree</span>(<span class="dt">text=</span><span class="st">'((a, b, c, d, e, f), (g, h, i, j));'</span>),</a>
<a class="sourceLine" id="cb21-2" data-line-number="2">                      <span class="kw">read.tree</span>(<span class="dt">text=</span><span class="st">'((a, b, c, d, e, i, j), (g, h, f));'</span>))</a></code></pre></div>
<pre><code>## [1] 3</code></pre>
<p>This distance is difficult to normalize, as its maximum value is difficult to calculate.</p>
<div id="information-theoretic-alternative" class="section level3">
<h3 class="hasAnchor">
<a href="#information-theoretic-alternative" class="anchor"></a>Information theoretic alternative</h3>
<p>In the matching split distance, <span class="math inline">\(m\)</span> represents a simple count of the number of shared taxa. An alternative is to measure the phylogenetic information content of the largest split consistent with <span class="math inline">\(S_1\)</span> and <span class="math inline">\(S_2\)</span>:</p>
<p><span class="math inline">\(m = \max\{h(A_1 \cap A_2 | B_1 \cap B_2), h(A_1 \cap B_2 | B_1 \cap A_2)\}\)</span></p>
<p>The most information-rich split consistent with</p>
<pre><code>ABCDEF  | GHIJ
ABCDE IJ|FGH  </code></pre>
<p>is <code>ABCDE | GH</code>, which contains</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1">TreeTools<span class="op">::</span><span class="kw"><a href="https://ms609.github.io/TreeTools/reference/SplitInformation.html">SplitInformation</a></span>(<span class="dv">5</span>, <span class="dv">2</span>)</a></code></pre></div>
<pre><code>## [1] 3.169925</code></pre>
<p>bits of phylogenetic information, which provides a similarity score for this pairing.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="kw"><a href="../reference/TreeDistance.html">MatchingSplitInfoDistance</a></span>(tree1, tree2)</a></code></pre></div>
<pre><code>## [1] 17.27586</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1"><span class="kw"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(MatchingSplitInfoDistance, tree1, tree2,</a>
<a class="sourceLine" id="cb28-2" data-line-number="2">                  <span class="dt">Plot=</span>TreeDistPlot, <span class="dt">leaveRoom=</span><span class="ot">FALSE</span>, <span class="dt">matchZeros =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-11-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-Bogdanowicz2012">
<p>Bogdanowicz, D., &amp; Giaro, K. (2012). Matching split distance for unrooted binary phylogenetic trees. <em>IEEE/ACM Transactions on Computational Biology and Bioinformatics</em>, <em>9</em>(1), 150â€“160. doi:<a href="https://doi.org/10.1109/TCBB.2011.48">10.1109/TCBB.2011.48</a></p>
</div>
<div id="ref-Bocker2013">
<p>BÃ¶cker, S., Canzar, S., &amp; Klau, G. W. (2013). The generalized Robinson-Foulds metric. In A. Darling &amp; J. Stoye (Eds.), <em>Algorithms in Bioinformatics. WABI 2013. Lecture Notes in Computer Science, vol 8126</em> (pp. 156â€“169). Berlin, Heidelberg: Springer. doi:<a href="https://doi.org/10.1007/978-3-642-40453-5_13">10.1007/978-3-642-40453-5_13</a></p>
</div>
<div id="ref-Nye2006">
<p>Nye, T. M. W., LiÃ², P., &amp; Gilks, W. R. (2006). A novel algorithm and web-based tool for comparing two alternative phylogenetic trees. <em>Bioinformatics</em>, <em>22</em>(1), 117â€“119. doi:<a href="https://doi.org/10.1093/bioinformatics/bti720">10.1093/bioinformatics/bti720</a></p>
</div>
<div id="ref-SmithDist">
<p>Smith, M. R. (2020). Information theoretic Generalized Robinson-Foulds metrics for comparing phylogenetic trees. <em>In Review</em>.</p>
</div>
<div id="ref-Steel2006">
<p>Steel, M. A., &amp; Penny, D. (2006). Maximum parsimony and the phylogenetic information in multistate characters. In V. A. Albert (Ed.), <em>Parsimony, phylogeny, and genomics</em> (pp. 163â€“178). Oxford: Oxford University Press.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#shared-phylogenetic-information">Shared phylogenetic information</a></li>
      <li><a href="#mutual-clustering-information">Mutual clustering information</a></li>
      <li><a href="#nye-et-al--tree-similarity-metric">Nye <em>et al.</em> tree similarity metric</a></li>
      <li><a href="#matching-split-distance">Matching Split Distance</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://community.dur.ac.uk/martin.smith/">Martin R. Smith</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '8bf44d312479af73720c5ca5fcc12eba',
    indexName: 'treedist',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
