<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Generalized Robinson-Foulds distances • TreeDist</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Generalized Robinson-Foulds distances">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TreeDist</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/Using-TreeDist.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Using-TreeDist.html">Using TreeDist to calculate tree similarity</a>
    </li>
    <li>
      <a href="../articles/using-distances.html">Contextualizing tree distances</a>
    </li>
    <li>
      <a href="../articles/tree-distances.html">Tree distance measures</a>
    </li>
    <li>
      <a href="../articles/Robinson-Foulds.html">Robinson-Foulds metric</a>
    </li>
    <li>
      <a href="../articles/Generalized-RF.html">Generalized Robinson-Foulds distances</a>
    </li>
    <li>
      <a href="../articles/information.html">Concepts of information</a>
    </li>
    <li>
      <a href="../articles/Tree-distance-metric-evaluation.html">Evaluating tree distance metrics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ms609/TreeDist">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Generalized Robinson-Foulds distances</h1>
                        <h4 class="author">Martin R. Smith</h4>
            
            <h4 class="date">2020-01-15</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ms609/TreeDist/blob/master/vignettes/Generalized-RF.Rmd"><code>vignettes/Generalized-RF.Rmd</code></a></small>
      <div class="hidden name"><code>Generalized-RF.Rmd</code></div>

    </div>

    
    
<p>This document outlines the similarity measures employed by the generalized Robinson-Foulds distances implemented in this package.</p>
<p>Before you read it, you may wish to refresh your memory of the <a href="Robinson-Foulds.html">Robinson-Foulds distance</a>, or how to <a href="Using-TreeDist.html">use the TreeDist package</a>. To understand the information-based metrics, it is useful to be familiar with <a href="information.html">concepts of information</a>.</p>
<div id="mutual-phylogenetic-information" class="section level2">
<h2 class="hasAnchor">
<a href="#mutual-phylogenetic-information" class="anchor"></a>Mutual phylogenetic information</h2>
<p>Under the mutual phylogenetic information tree distance measure <span class="citation">(Smith, 2019)</span>, pairs of partitions are assigned a similarity score that corresponds to their mutual phylogenetic information content.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(MutualPhylogeneticInfo, tree1, tree2, </a>
<a class="sourceLine" id="cb1-2" data-line-number="2">                  <span class="dt">Plot=</span>TreeDistPlot, <span class="dt">matchZeros =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-1-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="../reference/TreeDistance.html">MutualPhylogeneticInfo</a></span>(tree1, tree2)</a></code></pre></div>
<pre><code>## [1] 12.32188</code></pre>
<p>The phylogenetic information content of a partition is defined as the negative logarithm of the proportion of all bifurcating <em>n</em>-tip trees that contain that partition. Thus a more even bipartition, which is consistent with a smaller proportion of trees, contains more phylogenetic information.</p>
<p>The mutual phylogenetic information of two partitions is the information that they contain in common; that is, the sum of the information content of the two splits, minus the joint information (which is the negative logarithm of the proportion of trees that contain both partitions).</p>
<p>Consider the splits <code>ABCDEF:GHI</code> and <code>ABCDEFI:GH</code>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">'TreeTools'</span>)</a>
<a class="sourceLine" id="cb4-2" data-line-number="2">treesMatchingSplit &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(</a>
<a class="sourceLine" id="cb4-3" data-line-number="3">  <span class="dt">ABCDEF.GHI =</span> <span class="kw"><a href="https://ms609.github.io/TreeTools/reference/TreesMatchingSplit.html">TreesMatchingSplit</a></span>(<span class="dv">6</span>, <span class="dv">3</span>),</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">  <span class="dt">ABCDEFI.GH =</span> <span class="kw"><a href="https://ms609.github.io/TreeTools/reference/TreesMatchingSplit.html">TreesMatchingSplit</a></span>(<span class="dv">7</span>, <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb4-5" data-line-number="5">)</a>
<a class="sourceLine" id="cb4-6" data-line-number="6">treesMatchingSplit</a></code></pre></div>
<pre><code>## ABCDEF.GHI ABCDEFI.GH 
##       2835      10395</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">proportionMatchingSplit &lt;-<span class="st"> </span>treesMatchingSplit <span class="op">/</span><span class="st"> </span><span class="kw"><a href="https://ms609.github.io/TreeTools/reference/NRooted.html">NUnrooted</a></span>(<span class="dv">9</span>)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">proportionMatchingSplit</a></code></pre></div>
<pre><code>## ABCDEF.GHI ABCDEFI.GH 
## 0.02097902 0.07692308</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">splitInformation &lt;-<span class="st"> </span><span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(proportionMatchingSplit)</a>
<a class="sourceLine" id="cb8-2" data-line-number="2">splitInformation</a></code></pre></div>
<pre><code>## ABCDEF.GHI ABCDEFI.GH 
##   5.574909   3.700440</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">treesMatchingBoth &lt;-<span class="st"> </span><span class="kw"><a href="../reference/SplitMutualInformation.html">TreesConsistentWithTwoSplits</a></span>(<span class="dv">9</span>, <span class="dv">6</span>, <span class="dv">7</span>)</a>
<a class="sourceLine" id="cb10-2" data-line-number="2">jointInformation &lt;-<span class="st"> </span><span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(treesMatchingBoth <span class="op">/</span><span class="st"> </span><span class="kw"><a href="https://ms609.github.io/TreeTools/reference/NRooted.html">NUnrooted</a></span>(<span class="dv">9</span>))</a>
<a class="sourceLine" id="cb10-3" data-line-number="3"></a>
<a class="sourceLine" id="cb10-4" data-line-number="4">mutualInformation &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(splitInformation) <span class="op">-</span><span class="st"> </span>jointInformation</a>
<a class="sourceLine" id="cb10-5" data-line-number="5">mutualInformation</a></code></pre></div>
<pre><code>## [1] 2.115477</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="co"># Or more concisely:</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="kw"><a href="../reference/SplitMutualInformation.html">SplitMutualInformation</a></span>(<span class="dt">n =</span> <span class="dv">9</span>, <span class="dv">6</span>, <span class="dv">7</span>)</a></code></pre></div>
<pre><code>## [1] 2.115477</code></pre>
<!--#TODO: 1 != 1; need to normalize.-->
<div id="the-problem-with-arboreal-matchings" class="section level3">
<h3 class="hasAnchor">
<a href="#the-problem-with-arboreal-matchings" class="anchor"></a>The problem with arboreal matchings</h3>
<p>The mutual phylogenetic information is an arboreal similarity score: that is to say, it assigns zero similarity to partitions that cannot both occur on a single tree.<br>
This leads to problematic behaviour in certain cases: for example, when the position of two distant taxa is swapped, similarities within a tree can be obscured.<br>
The two trees below are identical but for the position of <em>A</em> and <em>J</em>, yet receive zero similarity under arboreal metrics.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(MutualPhylogeneticInfo, <span class="dt">matchZeros =</span> <span class="ot">FALSE</span>, AtoJ, AotJ,</a>
<a class="sourceLine" id="cb14-2" data-line-number="2">                  <span class="dt">Plot=</span>TreeDistPlot, <span class="dt">leaveRoom =</span> <span class="ot">FALSE</span>, <span class="dt">prune=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">5</span>, <span class="dv">18</span>))</a></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-3-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="mutual-clustering-information" class="section level2">
<h2 class="hasAnchor">
<a href="#mutual-clustering-information" class="anchor"></a>Mutual clustering information</h2>
<p>Clustering information <span class="citation">(Smith, 2019)</span> provides a non-arboreal information-based tree distance metric. As desired, this assigns a non-zero similarity to the case that received zero similarity under an arboreal matching:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(MutualClusteringInfo, AtoJ, AotJ, <span class="dt">matchZeros =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb15-2" data-line-number="2">                  <span class="dt">Plot=</span>TreeDistPlot, <span class="dt">leaveRoom =</span> <span class="ot">FALSE</span>, <span class="dt">prune=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">5</span>, <span class="dv">18</span>))</a></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="kw"><a href="../reference/TreeDistance.html">MutualClusteringInfo</a></span>(AtoJ, AotJ)</a></code></pre></div>
<pre><code>## [1] 13.48339</code></pre>
<p>Here, each partition is viewed as clustering the terminals into two groups. Consider a pair of partitions, Y<sub>1</sub> and Y<sub>2</sub>. If the partitions are identical, then if an observer knows which group a terminal is in Y<sub>1</sub>, then this gives all the information they need to know which group it is in in Y<sub>2</sub>. As the partitions become more different, the disposition of Y<sub>1</sub> gives less information about the configuration of Y<sub>2</sub>; the probability that a guess based on Y<sub>1</sub> will be correct decreases, and the mutual information decreases accordingly.</p>
<p>The mutual clustering information has the notable property that no pair of splits will be allocated zero similarity. As such, even a dissimilar matching (such as <code>HI:ABCDEFG</code> =&gt; <code>EI:ABCDFGH</code> below) is preferred to leaving a partition unpaired.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="kw"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(MutualClusteringInfo, tree1, tree2, </a>
<a class="sourceLine" id="cb18-2" data-line-number="2">                  <span class="dt">Plot=</span>TreeDistPlot, <span class="dt">leaveRoom=</span><span class="ot">FALSE</span>, <span class="dt">matchZeros =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-5-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="nye-et-al--tree-similarity-metric" class="section level2">
<h2 class="hasAnchor">
<a href="#nye-et-al--tree-similarity-metric" class="anchor"></a>Nye <em>et al.</em> tree similarity metric</h2>
<p>The Nye <em>et al</em>. <span class="citation">(2006)</span> tree similarity metric scores pairs by considering the elements held in common between subsets of each partition.</p>
<p>Consider a pair of partitions <code>ABCDEF:GHIJ</code> and <code>ABCDEIJ:FGH</code>. These can be aligned thus:</p>
<p><code>ABCDEF  : GHIJ</code> <code>ABCDE IJ:FGH</code></p>
<p>The first pair of subsets, <code>ABCDEF</code> and <code>ABCDEIJ</code>, have five elements in common (<code>ABCDE</code>), and together encompass eight elements (<code>ABCDEFIJ</code>). Their <em>subset score</em> is thus 5/8.</p>
<p>The second pair of subsets, <code>GHIJ</code> and <code>FGH</code>, have two elements (<code>GH</code>) in common, of the five total (<code>FGHIJ</code>), and hence receive a subset score of 2/5.</p>
<p>This split alignment then receives an <em>alignment score</em> corresponding to the lower of the two subset scores, 2/5.</p>
<p>We must now consider the other alignment of this pair of partitions,</p>
<p><code>ABCDEF  :     GHIJ</code> <code>FGH:ABCDE  IJ</code></p>
<p>This yields subset scores of 1/8 and 2/9, and thus has an alignment score of 1/8. This alignment gives a lower score than the other, so is disregarded. The pair of partitions is allocated a similarity score corresponding to the better alignment: 2/5.</p>
<p>As such, partitions that match exactly will receive a similarity score of 1, in a manner analogous to the Robinson-Foulds distance. (This is despite the fact that some partitions are more likely to match than others.) And it is not possible for a pair of partitions to receive a zero similarity score.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="kw"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(NyeTreeSimilarity, tree1, tree2, </a>
<a class="sourceLine" id="cb19-2" data-line-number="2">                  <span class="dt">Plot=</span>TreeDistPlot, <span class="dt">leaveRoom=</span><span class="ot">FALSE</span>, <span class="dt">matchZeros =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="kw"><a href="../reference/NyeTreeSimilarity.html">NyeTreeSimilarity</a></span>(tree1, tree2, <span class="dt">normalize=</span><span class="ot">FALSE</span>)</a></code></pre></div>
<pre><code>## [1] 3.5</code></pre>
<div id="jaccard-robinson-foulds-metric" class="section level3">
<h3 class="hasAnchor">
<a href="#jaccard-robinson-foulds-metric" class="anchor"></a>Jaccard-Robinson-Foulds metric</h3>
<p>Böcker <em>et al</em>. <span class="citation">(2013)</span> propose expressly prohibiting contradictory pairings, i.e. pairs of partitions that could never occur in the same tree. Such an ‘arboreal’ matching would preclude the pairing of <code>ABCFGHI:DE</code> with <code>ABCDFGH:EI</code>:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="kw"><a href="../reference/JaccardRobinsonFoulds.html">JaccardRobinsonFoulds</a></span>(tree1, tree2, <span class="dt">similarity =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] 6.333333</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"><span class="kw"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(JaccardRobinsonFoulds, tree1, tree2,</a>
<a class="sourceLine" id="cb24-2" data-line-number="2">                  <span class="dt">Plot=</span>TreeDistPlot, <span class="dt">leaveRoom=</span><span class="ot">FALSE</span>, <span class="dt">matchZeros =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-7-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Böcker <em>et al</em>. <span class="citation">(2013)</span> also suggest raising the partition similarity score defined above to an arbitrary exponent. The similarity score is also doubled. By this means, the metric can be made to converge to the Robinson-Foulds metric as the exponent grows towards infinity.</p>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="matching-split-distance" class="section level2">
<h2 class="hasAnchor">
<a href="#matching-split-distance" class="anchor"></a>Matching Split Distance</h2>
<p>Bogdanowicz &amp; Giaro <span class="citation">(2012)</span> propose an alternative distance, which they term the Matching Split Distance.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1"><span class="kw"><a href="../reference/MatchingSplitDistance.html">MatchingSplitDistance</a></span>(tree1, tree2)</a></code></pre></div>
<pre><code>## [1] 5</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="kw"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(MatchingSplitDistance, tree1, tree2,</a>
<a class="sourceLine" id="cb27-2" data-line-number="2">                  <span class="dt">Plot=</span>TreeDistPlot, <span class="dt">leaveRoom=</span><span class="ot">FALSE</span>, <span class="dt">matchZeros =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-9-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Note that the visualization shows the difference, rather than the similarity, between splits.</p>
<p>Similar to the Nye <em>et al</em>. similarity metric, this method compares the subsets implied by a pair of partitions. Here, the relevant quantity is the number of elements that must be moved from one subset to another in order to make the two partitions identical. With the pair of partitions</p>
<p><code>ABCDEF  : GHIJ</code> <code>ABCDE IJ:FGH</code></p>
<p>three elements (<em>F</em>, <em>I</em> and <em>J</em>) must be moved before the partitions are identical; as such, the pair of partitions are assigned a difference score (not a similarity score) of three.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1"><span class="kw"><a href="../reference/MatchingSplitDistance.html">MatchingSplitDistance</a></span>(<span class="kw">read.tree</span>(<span class="dt">text=</span><span class="st">'((a, b, c, d, e, f), (g, h, i, j));'</span>),</a>
<a class="sourceLine" id="cb28-2" data-line-number="2">                      <span class="kw">read.tree</span>(<span class="dt">text=</span><span class="st">'((a, b, c, d, e, i, j), (g, h, f));'</span>))</a></code></pre></div>
<pre><code>## [1] 3</code></pre>
<p>This distance is difficult to normalize, as its maximum value is difficult to calculate.</p>
<div id="information-theoretic-alternative" class="section level3">
<h3 class="hasAnchor">
<a href="#information-theoretic-alternative" class="anchor"></a>Information theoretic alternative</h3>
<p>An information-theoretic take on the Matching Split Distance <span class="citation">(Smith, 2019)</span> is provided by the function</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1"><span class="kw"><a href="../reference/TreeDistance.html">VariationOfMatchingSplitInfo</a></span>(tree1, tree2)</a></code></pre></div>
<pre><code>## [1] 17.27586</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1"><span class="kw"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(VariationOfMatchingSplitInfo, tree1, tree2,</a>
<a class="sourceLine" id="cb32-2" data-line-number="2">                  <span class="dt">Plot=</span>TreeDistPlot, <span class="dt">leaveRoom=</span><span class="ot">FALSE</span>, <span class="dt">matchZeros =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-11-1.png" width="576" style="display: block; margin: auto;"></p>
<p>This approach scores each pair of bipartitions according to the information content of the most information-rich bipartition that is consistent with both partitions in the pair.</p>
<p>The most information-rich bipartition consistent with</p>
<p><code>ABCDEF  : GHIJ</code> <code>ABCDE IJ:FGH</code></p>
<p>is <code>ABCDE : GH</code>, which contains</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1"><span class="kw"><a href="https://ms609.github.io/TreeTools/reference/SplitInformation.html">SplitInformation</a></span>(<span class="dv">5</span>, <span class="dv">2</span>)</a></code></pre></div>
<pre><code>## [1] 3.169925</code></pre>
<p>bits of information; this quantity represents the <em>similarity</em> score for this particular pairing.</p>
<p>This measure can therefore be normalized against the phylogenetic information content of the two input trees.</p>
</div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-Bogdanowicz2012">
<p>Bogdanowicz, D., &amp; Giaro, K. (2012). Matching split distance for unrooted binary phylogenetic trees. <em>IEEE/ACM Transactions on Computational Biology and Bioinformatics</em>, <em>9</em>(1), 150–160. doi:<a href="https://doi.org/10.1109/TCBB.2011.48">10.1109/TCBB.2011.48</a></p>
</div>
<div id="ref-Bocker2013">
<p>Böcker, S., Canzar, S., &amp; Klau, G. W. (2013). The generalized Robinson-Foulds metric. In A. Darling &amp; J. Stoye (Eds.), <em>Algorithms in Bioinformatics. WABI 2013. Lecture Notes in Computer Science, vol 8126</em> (pp. 156–169). Berlin, Heidelberg: Springer.</p>
</div>
<div id="ref-Nye2006">
<p>Nye, T. M. W., Liò, P., &amp; Gilks, W. R. (2006). A novel algorithm and web-based tool for comparing two alternative phylogenetic trees. <em>Bioinformatics</em>, <em>22</em>(1), 117–119. doi:<a href="https://doi.org/10.1093/bioinformatics/bti720">10.1093/bioinformatics/bti720</a></p>
</div>
<div id="ref-SmithDist">
<p>Smith, M. R. (2019). Information theoretic generalized robinson-foulds metrics for comparing phylogenetic trees. <em>Forthcoming</em>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#mutual-phylogenetic-information">Mutual phylogenetic information</a></li>
      <li><a href="#mutual-clustering-information">Mutual clustering information</a></li>
      <li><a href="#nye-et-al--tree-similarity-metric">Nye <em>et al.</em> tree similarity metric</a></li>
      <li><a href="#matching-split-distance">Matching Split Distance</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://community.dur.ac.uk/martin.smith/">Martin R. Smith</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '8bf44d312479af73720c5ca5fcc12eba',
    indexName: 'treedist',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
