<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Generalized Robinson-Foulds distances • TreeDist</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Generalized Robinson-Foulds distances">
<meta property="og:description" content="TreeDist">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-156666971-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-156666971-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TreeDist</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/Using-TreeDist.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Using-TreeDist.html">Using TreeDist to calculate tree similarity</a>
    </li>
    <li>
      <a href="../articles/using-distances.html">Contextualizing tree distances</a>
    </li>
    <li>
      <a href="../articles/Robinson-Foulds.html">Robinson-Foulds metric</a>
    </li>
    <li>
      <a href="../articles/Generalized-RF.html">Generalized Robinson-Foulds distances</a>
    </li>
    <li>
      <a href="../articles/information.html">Concepts of information</a>
    </li>
    <li>
      <a href="../articles/Tree-distance-metric-evaluation.html">Evaluating tree distance metrics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ms609/TreeDist">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Generalized Robinson-Foulds distances</h1>
                        <h4 class="author">Martin R. Smith</h4>
            
            <h4 class="date">2020-06-18</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ms609/TreeDist/blob/master/vignettes/Generalized-RF.Rmd"><code>vignettes/Generalized-RF.Rmd</code></a></small>
      <div class="hidden name"><code>Generalized-RF.Rmd</code></div>

    </div>

    
    
<p>This document outlines the similarity measures employed by the generalized Robinson-Foulds distances implemented in this package.</p>
<p>Generalized RF distances are introduced <a href="Robinson-Foulds.html">elsewhere</a>; before you read further, you may also wish to revisit how to <a href="Using-TreeDist.html">use the ‘TreeDist’ package</a>, and relevant <a href="information.html">principles of information theory</a>.</p>
<div id="shared-phylogenetic-information" class="section level2">
<h2 class="hasAnchor">
<a href="#shared-phylogenetic-information" class="anchor"></a>Shared phylogenetic information</h2>
<p>Under the shared phylogenetic information tree distance measure <span class="citation">(Smith, 2020)</span>, pairs of splits are assigned a similarity score that corresponds to the amount of phylogenetic information <span class="citation">(<em>sensu</em> Steel &amp; Penny, 2006)</span> that they share in common (see <a href="information.html">separate vignette</a>), a concept introduced (but not developed) by Nelson <span class="citation">(1979)</span>.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(<span class="no">SharedPhylogeneticInfo</span>, <span class="no">tree1</span>, <span class="no">tree2</span>,
                  <span class="kw">Plot</span> <span class="kw">=</span> <span class="no">TreeDistPlot</span>, <span class="kw">matchZeros</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-1-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="../reference/TreeDistance.html">SharedPhylogeneticInfo</a></span>(<span class="no">tree1</span>, <span class="no">tree2</span>)</pre></body></html></div>
<pre><code>## [1] 12.32188</code></pre>
<div id="conflicting-splits-can-nevertheless-be-instructive" class="section level3">
<h3 class="hasAnchor">
<a href="#conflicting-splits-can-nevertheless-be-instructive" class="anchor"></a>Conflicting splits can nevertheless be instructive</h3>
<p>Shared phylogenetic information assigns zero similarity to incompatible splits, i.e. those that cannot both occur on a single tree.<br>
This leads to problematic behaviour in certain cases: for example, swapping the position of two distant leaves (as with leaves ‘A’ and ‘J’ below) can disproportionately reduce similarity – in this example, to zero.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(<span class="no">SharedPhylogeneticInfo</span>, <span class="no">AtoJ</span>, <span class="no">swapAJ</span>,
                  <span class="kw">Plot</span> <span class="kw">=</span> <span class="no">TreeDistPlot</span>, <span class="kw">matchZeros</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">prune</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">5</span>, <span class="fl">18</span>))</pre></body></html></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-2-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="mutual-clustering-information" class="section level2">
<h2 class="hasAnchor">
<a href="#mutual-clustering-information" class="anchor"></a>Mutual clustering information</h2>
<p>Scoring each pair of splits according to their mutual clustering information <span class="citation">(Smith, 2020)</span> (see <a href="information.html">separate vignette</a>) results in a information-based tree distance metric that recognizes similarity in tree structure even when every possible pairing of splits conflicts:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(<span class="no">MutualClusteringInfo</span>, <span class="no">AtoJ</span>, <span class="no">swapAJ</span>,
                  <span class="kw">Plot</span> <span class="kw">=</span> <span class="no">TreeDistPlot</span>, <span class="kw">matchZeros</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">prune</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">5</span>, <span class="fl">18</span>))</pre></body></html></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-3-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="fu"><a href="../reference/TreeDistance.html">MutualClusteringInfo</a></span>(<span class="no">AtoJ</span>, <span class="no">swapAJ</span>)</pre></body></html></div>
<pre><code>## [1] 1.348339</code></pre>
<p>Because no pair of non-trivial splits has zero mutual clustering information, even a dissimilar pairing (such as <code>HI|ABCDEFG</code> =&gt; <code>EI|ABCDFGH</code> below) is (slightly) preferable to leaving a split unpaired.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(<span class="no">MutualClusteringInfo</span>, <span class="no">tree1</span>, <span class="no">tree2</span>,
                  <span class="kw">Plot</span> <span class="kw">=</span> <span class="no">TreeDistPlot</span>, <span class="kw">matchZeros</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="nye-et-al--tree-similarity-metric" class="section level2">
<h2 class="hasAnchor">
<a href="#nye-et-al--tree-similarity-metric" class="anchor"></a>Nye <em>et al.</em> tree similarity metric</h2>
<p>The Nye <em>et al</em>. <span class="citation">(2006)</span> tree similarity metric scores pairs by considering the elements held in common between subsets of each split.</p>
<p>Consider a pair of splits <code>ABCDEF|GHIJ</code> and <code>ABCDEIJ|FGH</code>. These can be aligned thus:</p>
<pre><code>ABCDEF  | GHIJ
ABCDE IJ|FGH</code></pre>
<p>The first pair of subsets, <code>ABCDEF</code> and <code>ABCDEIJ</code>, have five elements in common (<code>ABCDE</code>), and together encompass eight elements (<code>ABCDEFIJ</code>). Their <em>subset score</em> is thus 5/8.</p>
<p>The second pair of subsets, <code>GHIJ</code> and <code>FGH</code>, have two elements (<code>GH</code>) in common, of the five total (<code>FGHIJ</code>), and hence receive a subset score of 2/5.</p>
<p>This split alignment then receives an <em>alignment score</em> corresponding to the lower of the two subset scores, 2/5.</p>
<p>We must now consider the other alignment of this pair of splits,</p>
<pre><code>ABCDEF  |     GHIJ
     FGH|ABCDE  IJ</code></pre>
<p>This yields subset scores of 1/8 and 2/9, and thus has an alignment score of 1/8. This alignment gives a lower score than the other, so is disregarded. The pair of splits is allocated a similarity score corresponding to the better alignment: 2/5.</p>
<p>As such, splits that match exactly will receive a similarity score of 1, in a manner analogous to the Robinson-Foulds distance. (This is despite the fact that some splits are more likely to match than others.) It is not possible for a pair of splits to receive a zero similarity score.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(<span class="no">NyeTreeSimilarity</span>, <span class="no">tree1</span>, <span class="no">tree2</span>,
                  <span class="kw">Plot</span> <span class="kw">=</span> <span class="no">TreeDistPlot</span>, <span class="kw">matchZeros</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-5-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="../reference/NyeTreeSimilarity.html">NyeTreeSimilarity</a></span>(<span class="no">tree1</span>, <span class="no">tree2</span>, <span class="kw">normalize</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<pre><code>## [1] 3.5</code></pre>
<div id="jaccard-robinson-foulds-metric" class="section level3">
<h3 class="hasAnchor">
<a href="#jaccard-robinson-foulds-metric" class="anchor"></a>Jaccard-Robinson-Foulds metric</h3>
<p>Böcker <em>et al</em>. <span class="citation">(2013)</span> employ the same split similarity metric as Nye <em>et al.</em> <span class="citation">(2006)</span>, which they suggest ought to be raised to an arbitrary exponent in order to downweight the contribution of paired splits that are not identical. In order for the metric to converge to the Robinson-Foulds metric as the exponent grows towards infinity, the resulting score is then doubled.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu"><a href="../reference/JaccardRobinsonFoulds.html">JaccardRobinsonFoulds</a></span>(<span class="no">tree1</span>, <span class="no">tree2</span>, <span class="kw">k</span> <span class="kw">=</span> <span class="fl">1</span>)</pre></body></html></div>
<pre><code>## [1] 4</code></pre>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(<span class="no">JaccardRobinsonFoulds</span>, <span class="no">tree1</span>, <span class="no">tree2</span>,
                  <span class="kw">Plot</span> <span class="kw">=</span> <span class="no">TreeDistPlot</span>, <span class="kw">matchZeros</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="no">JRF2</span> <span class="kw">&lt;-</span> <span class="kw">function</span> (<span class="no">...</span>) <span class="fu"><a href="../reference/JaccardRobinsonFoulds.html">JaccardRobinsonFoulds</a></span>(<span class="kw">k</span> <span class="kw">=</span> <span class="fl">2</span>, <span class="no">...</span>)
<span class="fu">JRF2</span>(<span class="no">tree1</span>, <span class="no">tree2</span>)</pre></body></html></div>
<pre><code>## [1] 5.638889</code></pre>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(<span class="no">JRF2</span>, <span class="no">tree1</span>, <span class="no">tree2</span>,
                  <span class="kw">Plot</span> <span class="kw">=</span> <span class="no">TreeDistPlot</span>, <span class="kw">matchZeros</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-6-2.png" width="576" style="display: block; margin: auto;"></p>
<p>The figure below shows how the JRF distance between the two trees plotted above varies with the value of the exponent <em>k</em>, relative to the Nye <em>et al.</em> and Robinson-Foulds distances between the trees:</p>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-7-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The theoretical and practical performance of the JRF metric, and its speed of calculation, are best at lower values of <em>k</em> <span class="citation">(Smith, 2020)</span>, raising the question of whether an exponent is necessary.</p>
<p>Böcker <em>et al</em>. <span class="citation">(2013)</span> suggest that ‘reasonable’ matchings exhibit a property they term <em>arboreality</em>. Their definition of an arboreal matching supposes that trees are rooted, in which case each split corresponds to a clade. In an arboreal matching on a rooted tree, no pairing of splits conflicts with any other pairing of splits. Consider the case where splits ‘A’ and ‘B’ in tree 1 are paired respectively with splits ‘C’ and ‘D’ in tree 2. If ‘A’ is paired with ‘C’ and ‘B’ with ‘D’, then to avoid conflict:</p>
<ul>
<li><p>if A is nested within B, then C must be nested within D</p></li>
<li><p>if B is nested within A, then D must be nested within C</p></li>
<li><p>if A and B do not overlap, then C and D must not overlap.</p></li>
</ul>
<p>Equivalent statements for unrooted trees are a little harder to express.</p>
<p>Unfortunately, constructing arboreal matchings is NP-complete, making an optimal arboreal matching slow to find. A faster alternative is to prohibit pairings of contradictory splits, though distances generated under such an approach are theoretically less coherent and practically no more effective than those calculated when contradictory splits may be paired, so the only advantage of this approach is a slight increase in calculation speed <span class="citation">(Smith, 2020)</span>.</p>
</div>
</div>
<div id="matching-split-distance" class="section level2">
<h2 class="hasAnchor">
<a href="#matching-split-distance" class="anchor"></a>Matching Split Distance</h2>
<p>Bogdanowicz &amp; Giaro <span class="citation">(2012)</span> propose an alternative distance, which they term the Matching Split Distance. (This approach was independently proposed by Lin <em>et al.</em> <span class="citation">(2012)</span>.)</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="fu"><a href="../reference/MatchingSplitDistance.html">MatchingSplitDistance</a></span>(<span class="no">tree1</span>, <span class="no">tree2</span>)</pre></body></html></div>
<pre><code>## [1] 5</code></pre>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(<span class="no">MatchingSplitDistance</span>, <span class="no">tree1</span>, <span class="no">tree2</span>,
                  <span class="kw">Plot</span> <span class="kw">=</span> <span class="no">TreeDistPlot</span>, <span class="kw">matchZeros</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Note that the visualization shows the difference, rather than the similarity, between splits.</p>
<p>Similar to the Nye <em>et al</em>. similarity metric, this method compares the subsets implied by a pair of splits. Here, the relevant quantity is the number of elements that must be moved from one subset to another in order to make the two splits identical. With the pair of splits</p>
<pre><code>ABCDEF  | GHIJ
ABCDE IJ|FGH</code></pre>
<p>three leaves (‘F’, ‘I’ and ‘J’) must be moved before the splits are identical; as such, the pair of splits are assigned a difference score of three.<br>
Formally, where <span class="math inline">\(S_i\)</span> splits <span class="math inline">\(n\)</span> leaves into bipartitions <span class="math inline">\(A_i\)</span> and <span class="math inline">\(B_i\)</span>, the difference score is calculated by</p>
<p><span class="math inline">\(n - m\)</span></p>
<p>where <span class="math inline">\(m\)</span> counts the number of leaves that already match, and is defined as</p>
<p><span class="math inline">\(m = \max\{|A_1 \cap A_2| + |B_1 \cap B_2|, |A_1 \cap B_2| + |B_1 \cap A_2|\}\)</span></p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="fu"><a href="../reference/MatchingSplitDistance.html">MatchingSplitDistance</a></span>(<span class="fu">read.tree</span>(<span class="kw">text</span><span class="kw">=</span><span class="st">'((a, b, c, d, e, f), (g, h, i, j));'</span>),
                      <span class="fu">read.tree</span>(<span class="kw">text</span><span class="kw">=</span><span class="st">'((a, b, c, d, e, i, j), (g, h, f));'</span>))</pre></body></html></div>
<pre><code>## [1] 3</code></pre>
<p>This distance is difficult to normalize, as it is not easy to calculate its maximum value.</p>
<div id="information-theoretic-alternative" class="section level3">
<h3 class="hasAnchor">
<a href="#information-theoretic-alternative" class="anchor"></a>Information theoretic alternative</h3>
<p>In the matching split distance, <span class="math inline">\(m\)</span> represents a simple count of the number of shared taxa. An alternative is to measure the phylogenetic information content of the largest split consistent with <span class="math inline">\(S_1\)</span> and <span class="math inline">\(S_2\)</span>:</p>
<p><span class="math inline">\(m = \max\{h(A_1 \cap A_2 | B_1 \cap B_2), h(A_1 \cap B_2 | B_1 \cap A_2)\}\)</span></p>
<p>The most information-rich split consistent with</p>
<pre><code>ABCDEF  | GHIJ
ABCDE IJ|FGH  </code></pre>
<p>is <code>ABCDE | GH</code>, which contains</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="kw pkg">TreeTools</span><span class="kw ns">::</span><span class="fu"><a href="https://ms609.github.io/TreeTools/reference/SplitInformation.html">SplitInformation</a></span>(<span class="fl">5</span>, <span class="fl">2</span>)</pre></body></html></div>
<pre><code>## [1] 3.169925</code></pre>
<p>bits of phylogenetic information. This value can be used as a similarity score for this pairing of splits.</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="fu"><a href="../reference/TreeDistance.html">MatchingSplitInfoDistance</a></span>(<span class="no">tree1</span>, <span class="no">tree2</span>)</pre></body></html></div>
<pre><code>## [1] 17.27586</code></pre>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="fu"><a href="../reference/VisualizeMatching.html">VisualizeMatching</a></span>(<span class="no">MatchingSplitInfoDistance</span>, <span class="no">tree1</span>, <span class="no">tree2</span>,
                  <span class="kw">Plot</span> <span class="kw">=</span> <span class="no">TreeDistPlot</span>, <span class="kw">matchZeros</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<p><img src="Generalized-RF_files/figure-html/unnamed-chunk-11-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-Bogdanowicz2012">
<p>Bogdanowicz, D., &amp; Giaro, K. (2012). Matching split distance for unrooted binary phylogenetic trees. <em>IEEE/ACM Transactions on Computational Biology and Bioinformatics</em>, <em>9</em>(1), 150–160. doi:<a href="https://doi.org/10.1109/TCBB.2011.48">10.1109/TCBB.2011.48</a></p>
</div>
<div id="ref-Bocker2013">
<p>Böcker, S., Canzar, S., &amp; Klau, G. W. (2013). The generalized Robinson-Foulds metric. In A. Darling &amp; J. Stoye (Eds.), <em>Algorithms in Bioinformatics. WABI 2013. Lecture Notes in Computer Science, vol 8126</em> (pp. 156–169). Berlin, Heidelberg: Springer. doi:<a href="https://doi.org/10.1007/978-3-642-40453-5_13">10.1007/978-3-642-40453-5_13</a></p>
</div>
<div id="ref-Lin2012">
<p>Lin, Y., Rajan, V., &amp; Moret, B. M. E. (2012). A metric for phylogenetic trees based on matching. <em>IEEE/ACM Transactions on Computational Biology and Bioinformatics</em>, <em>4</em>(9), 1014–1022. doi:<a href="https://doi.org/10.1109/TCBB.2011.157">10.1109/TCBB.2011.157</a></p>
</div>
<div id="ref-Nelson1979">
<p>Nelson, G. (1979). Cladistic analysis and synthesis: Principles and definitions, with a historical note on Adanson’s <em>Familles des Plantes</em> (1763–1764). <em>Systematic Biology</em>, <em>28</em>(1), 1–21. doi:<a href="https://doi.org/10.1093/sysbio/28.1.1">10.1093/sysbio/28.1.1</a></p>
</div>
<div id="ref-Nye2006">
<p>Nye, T. M. W., Liò, P., &amp; Gilks, W. R. (2006). A novel algorithm and web-based tool for comparing two alternative phylogenetic trees. <em>Bioinformatics</em>, <em>22</em>(1), 117–119. doi:<a href="https://doi.org/10.1093/bioinformatics/bti720">10.1093/bioinformatics/bti720</a></p>
</div>
<div id="ref-SmithDist">
<p>Smith, M. R. (2020). Information theoretic Generalized Robinson-Foulds metrics for comparing phylogenetic trees. <em>In Review</em>.</p>
</div>
<div id="ref-Steel2006">
<p>Steel, M. A., &amp; Penny, D. (2006). Maximum parsimony and the phylogenetic information in multistate characters. In V. A. Albert (Ed.), <em>Parsimony, phylogeny, and genomics</em> (pp. 163–178). Oxford: Oxford University Press.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://community.dur.ac.uk/martin.smith/">Martin R. Smith</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '8bf44d312479af73720c5ca5fcc12eba',
    indexName: 'treedist',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
